// @rarible/sdk-transaction/build/immutablex/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockchainImmutableXTransaction = void 0;
const tslib_1 = require("tslib");
const api_client_1 = require("@rarible/api-client");
class BlockchainImmutableXTransaction {
    constructor(transaction) {
        this.transaction = transaction;
        this.blockchain = api_client_1.Blockchain.IMMUTABLEX;
    }
    hash() {
        var _a;
        return ((_a = this.transaction) !== null && _a !== void 0 ? _a : "") + "";
    }
    wait() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return {
                blockchain: this.blockchain,
                hash: this.hash(),
            };
        });
    }
    getTxLink() {
        if (!this.transaction) {
            return "";
        }
        return "https://immutascan.io/tx/" + this.transaction;
    }
    get isEmpty() {
        return this.transaction === undefined;
    }
}
exports.BlockchainImmutableXTransaction = BlockchainImmutableXTransaction;
