// @rarible/eclipse-sdk/build/sdk/order/cancel-sell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cancelSell = void 0;
const tslib_1 = require("tslib");
const web3_js_1 = require("@solana/web3.js");
const web3_js_2 = require("@solana/web3.js");
const spl_token_1 = require("@solana/spl-token");
const marketplace_program_1 = require("../core/marketplace-program");
const utils_1 = require("../utils");
function cancelSell(request) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const marketProgram = (0, marketplace_program_1.getProgramInstanceRaribleMarketplace)(request.connection);
        const initializer = request.signer.publicKey.toString();
        const order = yield (0, utils_1.fetchOrderByAddress)(request.connection, request.orderAddress.toString());
        if (!order) {
            throw new Error(`Can't find order by given address: ${request.orderAddress.toString()}`);
        }
        const { nftMint } = order;
        const nftTokenProgram = yield (0, utils_1.getTokenProgramFromMint)(request.connection, nftMint.toString());
        if (!nftTokenProgram) {
            throw new Error(`Can't find token program for mint ${nftMint}`);
        }
        const initializerNftTa = (0, utils_1.getAtaAddress)(nftMint.toString(), initializer, nftTokenProgram.toString());
        const nftProgram = yield (0, utils_1.getNftProgramFromMint)(request.connection, nftMint.toString());
        const eventAuthority = (0, utils_1.getEventAuthority)();
        const remainingAccounts = yield (0, utils_1.getRemainingAccountsForMint)(request.connection, nftMint.toString(), request.extraAccountParams);
        const instruction = yield marketProgram.methods
            .cancelListing()
            .accountsStrict({
            initializer: request.signer.publicKey,
            market: order.market,
            nftMint,
            order: request.orderAddress,
            initializerNftTa,
            nftProgram: nftProgram !== null && nftProgram !== void 0 ? nftProgram : web3_js_2.PublicKey.default,
            nftTokenProgram,
            sysvarInstructions: web3_js_2.SYSVAR_INSTRUCTIONS_PUBKEY,
            associatedTokenProgram: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
            systemProgram: web3_js_2.SystemProgram.programId,
            program: utils_1.MARKETPLACE_PROGRAM_ID,
            eventAuthority,
        })
            .remainingAccounts(remainingAccounts)
            .instruction();
        const instructions = [];
        instructions.push(web3_js_1.ComputeBudgetProgram.setComputeUnitLimit({
            units: 850000,
        }));
        instructions.push(instruction);
        return {
            instructions,
            signers: [],
        };
    });
}
exports.cancelSell = cancelSell;
