// @rarible/eclipse-sdk/build/sdk/order/initialize-market.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeMarket = void 0;
const tslib_1 = require("tslib");
const web3_js_1 = require("@solana/web3.js");
const anchor_1 = require("@coral-xyz/anchor");
const marketplace_program_1 = require("../core/marketplace-program");
const utils_1 = require("../utils");
function initializeMarket(request) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const marketProgram = (0, marketplace_program_1.getProgramInstanceRaribleMarketplace)(request.connection);
        const market = (0, utils_1.getMarketPda)(request.marketIdentifier.toString());
        const eventAuthority = (0, utils_1.getEventAuthority)();
        const instruction = yield marketProgram.methods
            .initMarket({
            feeBps: new anchor_1.BN(request.feeBps.toString()),
            feeRecipient: request.feeRecipient,
            feeBps2: new anchor_1.BN(request.feeBps2.toString()),
            feeRecipient2: request.feeRecipient2,
        })
            .accountsStrict({
            initializer: request.signer.publicKey,
            marketIdentifier: request.marketIdentifier,
            market,
            systemProgram: web3_js_1.SystemProgram.programId,
            program: utils_1.MARKETPLACE_PROGRAM_ID,
            eventAuthority,
        })
            .instruction();
        const instructions = [instruction];
        return {
            instructions,
            signers: [],
        };
    });
}
exports.initializeMarket = initializeMarket;
