// @rarible/sdk/build/sdk-blockchains/tezos/token-id.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TezosTokenId = void 0;
const tslib_1 = require("tslib");
// eslint-disable-next-line camelcase
const tezos_sdk_1 = require("@rarible/tezos-sdk");
const types_1 = require("@rarible/types");
const common_1 = require("./common");
class TezosTokenId {
    constructor(provider) {
        this.provider = provider;
        this.generateTokenId = this.generateTokenId.bind(this);
    }
    generateTokenId(_a) {
        return tslib_1.__awaiter(this, arguments, void 0, function* ({ collection }) {
            const tokenId = yield (0, tezos_sdk_1.get_next_token_id)((0, common_1.getRequiredProvider)(this.provider), (0, common_1.convertFromContractAddress)(collection));
            return {
                tokenId: tokenId.toString(),
                signature: {
                    v: 0,
                    r: (0, types_1.toBinary)("0"),
                    s: (0, types_1.toBinary)("0"),
                },
            };
        });
    }
}
exports.TezosTokenId = TezosTokenId;
