// @rarible/types/esm/blockchains/union/enum/domain.js
export var BlockchainLayer1Enum;
(function (BlockchainLayer1Enum) {
    BlockchainLayer1Enum["ETHEREUM"] = "ETHEREUM";
    BlockchainLayer1Enum["FLOW"] = "FLOW";
    BlockchainLayer1Enum["TEZOS"] = "TEZOS";
    BlockchainLayer1Enum["SOLANA"] = "SOLANA";
    BlockchainLayer1Enum["APTOS"] = "APTOS";
})(BlockchainLayer1Enum || (BlockchainLayer1Enum = {}));
export const blockchainsLayer1 = Object.values(BlockchainLayer1Enum);
export var BlockchainEnum;
(function (BlockchainEnum) {
    BlockchainEnum["ETHEREUM"] = "ETHEREUM";
    BlockchainEnum["FLOW"] = "FLOW";
    BlockchainEnum["TEZOS"] = "TEZOS";
    BlockchainEnum["POLYGON"] = "POLYGON";
    BlockchainEnum["IMMUTABLEX"] = "IMMUTABLEX";
    BlockchainEnum["SOLANA"] = "SOLANA";
    BlockchainEnum["APTOS"] = "APTOS";
    BlockchainEnum["ECLIPSE"] = "ECLIPSE";
    BlockchainEnum["MANTLE"] = "MANTLE";
    BlockchainEnum["ZKSYNC"] = "ZKSYNC";
    BlockchainEnum["RARI"] = "RARI";
    BlockchainEnum["ASTARZKEVM"] = "ASTARZKEVM";
    BlockchainEnum["KROMA"] = "KROMA";
    BlockchainEnum["BASE"] = "BASE";
    BlockchainEnum["ARBITRUM"] = "ARBITRUM";
    BlockchainEnum["LIGHTLINK"] = "LIGHTLINK";
    BlockchainEnum["CELO"] = "CELO";
    BlockchainEnum["XAI"] = "XAI";
    BlockchainEnum["CHILIZ"] = "CHILIZ";
    BlockchainEnum["MOONBEAM"] = "MOONBEAM";
    BlockchainEnum["PALM"] = "PALM";
    BlockchainEnum["ETHERLINK"] = "ETHERLINK";
    BlockchainEnum["LISK"] = "LISK";
    BlockchainEnum["OASYS"] = "OASYS";
    BlockchainEnum["SAAKURU"] = "SAAKURU";
    BlockchainEnum["ALEPHZERO"] = "ALEPHZERO";
    BlockchainEnum["SHAPE"] = "SHAPE";
    BlockchainEnum["BERACHAIN"] = "BERACHAIN";
    BlockchainEnum["TELOS"] = "TELOS";
    BlockchainEnum["MATCH"] = "MATCH";
    BlockchainEnum["ABSTRACT"] = "ABSTRACT";
    BlockchainEnum["HEDERAEVM"] = "HEDERAEVM";
    BlockchainEnum["VICTION"] = "VICTION";
    BlockchainEnum["GOAT"] = "GOAT";
})(BlockchainEnum || (BlockchainEnum = {}));
export const blockchains = Object.values(BlockchainEnum);
export const evmBlockchains = [
    BlockchainEnum.POLYGON,
    BlockchainEnum.IMMUTABLEX,
    BlockchainEnum.POLYGON,
    BlockchainEnum.MANTLE,
    BlockchainEnum.ZKSYNC,
    BlockchainEnum.RARI,
    BlockchainEnum.ASTARZKEVM,
    BlockchainEnum.BASE,
    BlockchainEnum.ARBITRUM,
    BlockchainEnum.LIGHTLINK,
    BlockchainEnum.CELO,
    BlockchainEnum.KROMA,
    BlockchainEnum.XAI,
    BlockchainEnum.CHILIZ,
    BlockchainEnum.MOONBEAM,
    BlockchainEnum.PALM,
    BlockchainEnum.ETHERLINK,
    BlockchainEnum.LISK,
    BlockchainEnum.OASYS,
    BlockchainEnum.SAAKURU,
    BlockchainEnum.ALEPHZERO,
    BlockchainEnum.SHAPE,
    BlockchainEnum.BERACHAIN,
    BlockchainEnum.TELOS,
    BlockchainEnum.MATCH,
    BlockchainEnum.ABSTRACT,
    BlockchainEnum.HEDERAEVM,
    BlockchainEnum.VICTION,
    BlockchainEnum.GOAT,
];
export function isEVMBlockchain(blockchain) {
    return evmBlockchains.includes(blockchain);
}
export const L1BlockchainByBlockchainDictionary = {
    [BlockchainEnum.ETHEREUM]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.POLYGON]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.MANTLE]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.RARI]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.IMMUTABLEX]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.ZKSYNC]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.ASTARZKEVM]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.ARBITRUM]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.LIGHTLINK]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.CELO]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.KROMA]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.XAI]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.CHILIZ]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.BASE]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.MOONBEAM]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.PALM]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.ETHERLINK]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.LISK]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.OASYS]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.SAAKURU]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.FLOW]: BlockchainLayer1Enum.FLOW,
    [BlockchainEnum.SOLANA]: BlockchainLayer1Enum.SOLANA,
    [BlockchainEnum.ECLIPSE]: BlockchainLayer1Enum.SOLANA,
    [BlockchainEnum.APTOS]: BlockchainLayer1Enum.APTOS,
    [BlockchainEnum.TEZOS]: BlockchainLayer1Enum.TEZOS,
    [BlockchainEnum.ALEPHZERO]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.SHAPE]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.BERACHAIN]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.TELOS]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.MATCH]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.ABSTRACT]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.HEDERAEVM]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.VICTION]: BlockchainLayer1Enum.ETHEREUM,
    [BlockchainEnum.GOAT]: BlockchainLayer1Enum.ETHEREUM,
};
