// @rarible/logger/build/remote/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteLogger = void 0;
var tslib_1 = require("tslib");
var batcher_1 = require("../utils/batcher");
var domain_1 = require("../domain");
var get_loggable_message_1 = require("../utils/get-loggable-message");
var RemoteLogger = /** @class */ (function () {
    function RemoteLogger(handler, config) {
        this.config = (0, tslib_1.__assign)((0, tslib_1.__assign)({}, defaultConfig), config);
        this.batchManager = new batcher_1.Batcher(this.config.dropBatchInterval, handler);
    }
    RemoteLogger.prototype.debug = function () {
        var params = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            params[_i] = arguments[_i];
        }
        this.log.apply(this, (0, tslib_1.__spreadArray)([domain_1.LogLevel.DEBUG], params, true)).then();
    };
    RemoteLogger.prototype.error = function () {
        var params = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            params[_i] = arguments[_i];
        }
        this.log.apply(this, (0, tslib_1.__spreadArray)([domain_1.LogLevel.ERROR], params, true)).then();
    };
    RemoteLogger.prototype.info = function () {
        var params = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            params[_i] = arguments[_i];
        }
        this.log.apply(this, (0, tslib_1.__spreadArray)([domain_1.LogLevel.INFO], params, true)).then();
    };
    RemoteLogger.prototype.trace = function () {
        var params = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            params[_i] = arguments[_i];
        }
        this.log.apply(this, (0, tslib_1.__spreadArray)([domain_1.LogLevel.TRACE], params, true)).then();
    };
    RemoteLogger.prototype.warn = function () {
        var params = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            params[_i] = arguments[_i];
        }
        this.log.apply(this, (0, tslib_1.__spreadArray)([domain_1.LogLevel.WARN], params, true)).then();
    };
    RemoteLogger.prototype.raw = function (data) {
        this.rawAsync(data).then();
    };
    RemoteLogger.prototype.log = function (level) {
        var params = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            params[_i - 1] = arguments[_i];
        }
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.rawAsync({
                            level: level,
                            message: get_loggable_message_1.getLoggableMessage.apply(void 0, (0, tslib_1.__spreadArray)([this.config.maxByteSize], params, false)),
                        })];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    RemoteLogger.prototype.rawAsync = function (data) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var finalData, e_1;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this.withContext(data)];
                    case 1:
                        finalData = _a.sent();
                        this.batchManager.add(finalData);
                        return [3 /*break*/, 3];
                    case 2:
                        e_1 = _a.sent();
                        console.error("Cannot connect to ELK", e_1);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    RemoteLogger.prototype.withContext = function (values) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var context;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.config.initialContext];
                    case 1:
                        context = _a.sent();
                        return [2 /*return*/, (0, tslib_1.__assign)((0, tslib_1.__assign)((0, tslib_1.__assign)({}, context), this.config.context), values)];
                }
            });
        });
    };
    return RemoteLogger;
}());
exports.RemoteLogger = RemoteLogger;
var defaultConfig = {
    maxByteSize: 10240,
    dropBatchInterval: 3000,
    context: {},
    initialContext: Promise.resolve({}),
};
