// @rarible/aptos-sdk/build/balance/balance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AptosBalance = void 0;
const tslib_1 = require("tslib");
const ts_sdk_1 = require("@aptos-labs/ts-sdk");
const utils_1 = require("@rarible/utils");
const types_1 = require("@rarible/types");
class AptosBalance {
    constructor(aptos) {
        this.aptos = aptos;
    }
    getAptosBalance(_a) {
        return tslib_1.__awaiter(this, arguments, void 0, function* ({ address }) {
            var _b;
            const [balance] = yield this.aptos.getAccountCoinsData({
                accountAddress: address,
                options: {
                    where: {
                        asset_type: {
                            _eq: ts_sdk_1.APTOS_COIN,
                        },
                    },
                },
            });
            if (!balance) {
                return (0, types_1.toBigNumber)("0");
            }
            return (0, types_1.toBigNumber)((0, utils_1.toBn)(balance.amount)
                .div((0, utils_1.toBn)(10).pow(((_b = balance.metadata) === null || _b === void 0 ? void 0 : _b.decimals) || 8))
                .toFixed());
        });
    }
}
exports.AptosBalance = AptosBalance;
