// @rarible/ethereum-api-client/build/apis/NftCollectionControllerApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * OpenAPI definition
 * # Exciting News  Hey there, amazing developers! 👋  We\'ve got some fantastic news to share that will make your experience with our product even better. We\'re thrilled to announce the implementation of API-Keys, which brings a new level of control and convenience to your development process. So, let\'s dive right into the details! ✨  ## API-Keys for Enhanced Functionality 🔑  Starting now, we have introduced API-Keys to streamline your interactions with our methods. You might have noticed a slight change while making requests – now, all you need to do is include the parameter `X-API-KEY` in the Header of your request. Simple as that!  ## Get Your API-Key in a Snap! ⚡️  Securing your API-Key is a breeze. We\'ve made the process super user-friendly to ensure you can get started quickly. Just head over to our website and fill out a simple form. Once you\'ve done that, your shiny new API-Key will be delivered straight to your email inbox. Easy peasy! 📧  [Get Your API-Key Here](https://api.rarible.org/registration) [Configure SDK with API-key](https://github.com/rarible/sdk#api-querying)  ## Unlock the Power of the Rarible Protocol 🔓  As passionate developers, we know you\'re always hungry for knowledge and eager to explore new frontiers. That\'s why we invite you to discover the incredible world of the Rarible Protocol. By visiting our dedicated protocol page, you\'ll gain access to a treasure trove of useful information, tips, and insights that will elevate your development skills to new heights. 🚀  [Explore the Rarible Protocol](https://rarible.org)  So, buckle up and get ready for an enhanced development journey with our API-Keys. We\'re excited to see what you\'ll create using our revamped system! 💪🚀  Stay curious, keep innovating, and happy coding! ✨  [P.S. Join our Discord Server to stay up to date and ask questions](https://discord.gg/rarifoundation)
 *
 * The version of the OpenAPI document: v0.1
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NftCollectionControllerApi = void 0;
const runtime = __importStar(require("../runtime"));
/**
 * no description
 */
class NftCollectionControllerApi extends runtime.BaseAPI {
    /**
     * Returns next available tokenId for minter
     */
    generateNftTokenIdRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.collection === null || requestParameters.collection === undefined) {
                throw new runtime.RequiredError('collection', 'Required parameter requestParameters.collection was null or undefined when calling generateNftTokenId.');
            }
            if (requestParameters.minter === null || requestParameters.minter === undefined) {
                throw new runtime.RequiredError('minter', 'Required parameter requestParameters.minter was null or undefined when calling generateNftTokenId.');
            }
            const queryParameters = {};
            if (requestParameters.minter !== undefined) {
                queryParameters['minter'] = requestParameters.minter;
            }
            const headerParameters = {};
            if (this.configuration && this.configuration.apiKey) {
                headerParameters["X-API-KEY"] = this.configuration.apiKey("X-API-KEY"); // ApiKeyAuth authentication
            }
            const response = yield this.request({
                path: `/v0.1/nft/collections/{collection}/generate_token_id`.replace(`{${"collection"}}`, encodeURIComponent(String(requestParameters.collection))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            if (response.status === 200) {
                return {
                    status: 200,
                    value: yield response.json()
                };
            }
            if (response.status === 400) {
                return {
                    status: 400,
                    value: yield response.json()
                };
            }
            if (response.status === 500) {
                return {
                    status: 500,
                    value: yield response.json()
                };
            }
            throw response;
        });
    }
    /**
     * Returns next available tokenId for minter
     */
    generateNftTokenId(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.generateNftTokenIdRaw(requestParameters);
            if (response.status === 200) {
                return response.value;
            }
            throw response;
        });
    }
    /**
     * Returns collection by address
     */
    getCollectionMetaRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.collection === null || requestParameters.collection === undefined) {
                throw new runtime.RequiredError('collection', 'Required parameter requestParameters.collection was null or undefined when calling getCollectionMeta.');
            }
            const queryParameters = {};
            const headerParameters = {};
            if (this.configuration && this.configuration.apiKey) {
                headerParameters["X-API-KEY"] = this.configuration.apiKey("X-API-KEY"); // ApiKeyAuth authentication
            }
            const response = yield this.request({
                path: `/v0.1/nft/collections/{collection}/meta`.replace(`{${"collection"}}`, encodeURIComponent(String(requestParameters.collection))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            if (response.status === 200) {
                return {
                    status: 200,
                    value: yield response.json()
                };
            }
            if (response.status === 400) {
                return {
                    status: 400,
                    value: yield response.json()
                };
            }
            if (response.status === 500) {
                return {
                    status: 500,
                    value: yield response.json()
                };
            }
            throw response;
        });
    }
    /**
     * Returns collection by address
     */
    getCollectionMeta(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getCollectionMetaRaw(requestParameters);
            if (response.status === 200) {
                return response.value;
            }
            throw response;
        });
    }
    /**
     * Returns collection by address
     */
    getNftCollectionByIdRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.collection === null || requestParameters.collection === undefined) {
                throw new runtime.RequiredError('collection', 'Required parameter requestParameters.collection was null or undefined when calling getNftCollectionById.');
            }
            const queryParameters = {};
            const headerParameters = {};
            if (this.configuration && this.configuration.apiKey) {
                headerParameters["X-API-KEY"] = this.configuration.apiKey("X-API-KEY"); // ApiKeyAuth authentication
            }
            const response = yield this.request({
                path: `/v0.1/nft/collections/{collection}`.replace(`{${"collection"}}`, encodeURIComponent(String(requestParameters.collection))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            if (response.status === 200) {
                return {
                    status: 200,
                    value: yield response.json()
                };
            }
            if (response.status === 404) {
                return {
                    status: 404,
                    value: yield response.json()
                };
            }
            if (response.status === 400) {
                return {
                    status: 400,
                    value: yield response.json()
                };
            }
            if (response.status === 500) {
                return {
                    status: 500,
                    value: yield response.json()
                };
            }
            throw response;
        });
    }
    /**
     * Returns collection by address
     */
    getNftCollectionById(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getNftCollectionByIdRaw(requestParameters);
            if (response.status === 200) {
                return response.value;
            }
            throw response;
        });
    }
    /**
     * Returns collections by ids
     */
    getNftCollectionsByIdsRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.collectionsByIdRequest === null || requestParameters.collectionsByIdRequest === undefined) {
                throw new runtime.RequiredError('collectionsByIdRequest', 'Required parameter requestParameters.collectionsByIdRequest was null or undefined when calling getNftCollectionsByIds.');
            }
            const queryParameters = {};
            const headerParameters = {};
            headerParameters['Content-Type'] = 'application/json';
            if (this.configuration && this.configuration.apiKey) {
                headerParameters["X-API-KEY"] = this.configuration.apiKey("X-API-KEY"); // ApiKeyAuth authentication
            }
            const response = yield this.request({
                path: `/v0.1/nft/collections/byId`,
                method: 'POST',
                headers: headerParameters,
                query: queryParameters,
                body: requestParameters.collectionsByIdRequest,
            });
            if (response.status === 200) {
                return {
                    status: 200,
                    value: yield response.json()
                };
            }
            if (response.status === 400) {
                return {
                    status: 400,
                    value: yield response.json()
                };
            }
            if (response.status === 500) {
                return {
                    status: 500,
                    value: yield response.json()
                };
            }
            throw response;
        });
    }
    /**
     * Returns collections by ids
     */
    getNftCollectionsByIds(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getNftCollectionsByIdsRaw(requestParameters);
            if (response.status === 200) {
                return response.value;
            }
            throw response;
        });
    }
    /**
     * Returns all NFT collections
     */
    searchNftAllCollectionsRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            const queryParameters = {};
            if (requestParameters.continuation !== undefined) {
                queryParameters['continuation'] = requestParameters.continuation;
            }
            if (requestParameters.size !== undefined) {
                queryParameters['size'] = requestParameters.size;
            }
            const headerParameters = {};
            if (this.configuration && this.configuration.apiKey) {
                headerParameters["X-API-KEY"] = this.configuration.apiKey("X-API-KEY"); // ApiKeyAuth authentication
            }
            const response = yield this.request({
                path: `/v0.1/nft/collections/all`,
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            if (response.status === 200) {
                return {
                    status: 200,
                    value: yield response.json()
                };
            }
            if (response.status === 400) {
                return {
                    status: 400,
                    value: yield response.json()
                };
            }
            if (response.status === 500) {
                return {
                    status: 500,
                    value: yield response.json()
                };
            }
            throw response;
        });
    }
    /**
     * Returns all NFT collections
     */
    searchNftAllCollections(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.searchNftAllCollectionsRaw(requestParameters);
            if (response.status === 200) {
                return response.value;
            }
            throw response;
        });
    }
    /**
     * Returns collection by owner
     */
    searchNftCollectionsByOwnerRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.owner === null || requestParameters.owner === undefined) {
                throw new runtime.RequiredError('owner', 'Required parameter requestParameters.owner was null or undefined when calling searchNftCollectionsByOwner.');
            }
            const queryParameters = {};
            if (requestParameters.owner !== undefined) {
                queryParameters['owner'] = requestParameters.owner;
            }
            if (requestParameters.continuation !== undefined) {
                queryParameters['continuation'] = requestParameters.continuation;
            }
            if (requestParameters.size !== undefined) {
                queryParameters['size'] = requestParameters.size;
            }
            const headerParameters = {};
            if (this.configuration && this.configuration.apiKey) {
                headerParameters["X-API-KEY"] = this.configuration.apiKey("X-API-KEY"); // ApiKeyAuth authentication
            }
            const response = yield this.request({
                path: `/v0.1/nft/collections/byOwner`,
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            if (response.status === 200) {
                return {
                    status: 200,
                    value: yield response.json()
                };
            }
            if (response.status === 400) {
                return {
                    status: 400,
                    value: yield response.json()
                };
            }
            if (response.status === 500) {
                return {
                    status: 500,
                    value: yield response.json()
                };
            }
            throw response;
        });
    }
    /**
     * Returns collection by owner
     */
    searchNftCollectionsByOwner(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.searchNftCollectionsByOwnerRaw(requestParameters);
            if (response.status === 200) {
                return response.value;
            }
            throw response;
        });
    }
}
exports.NftCollectionControllerApi = NftCollectionControllerApi;
