// @rarible/protocol-ethereum-sdk/build/nft/contracts/erc1155/deploy/rarible-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.erc1155FactoryABI = exports.createTestErc1155RaribleFactoryContract = exports.createErc1155FactoryContract = void 0;
const contracts_1 = require("../../../../common/contracts");
function createErc1155FactoryContract(ethereum, address) {
    return ethereum.createContract(exports.erc1155FactoryABI, address);
}
exports.createErc1155FactoryContract = createErc1155FactoryContract;
function createTestErc1155RaribleFactoryContract(web3, address) {
    return new web3.eth.Contract(exports.erc1155FactoryABI, address, contracts_1.NumberDataFormat);
}
exports.createTestErc1155RaribleFactoryContract = createTestErc1155RaribleFactoryContract;
exports.erc1155FactoryABI = [
    {
        inputs: [
            {
                internalType: "address",
                name: "_beacon",
                type: "address",
            },
            {
                internalType: "address",
                name: "_transferProxy",
                type: "address",
            },
            {
                internalType: "address",
                name: "_lazyTransferProxy",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "proxy",
                type: "address",
            },
        ],
        name: "Create1155RaribleProxy",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "previousOwner",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "newOwner",
                type: "address",
            },
        ],
        name: "OwnershipTransferred",
        type: "event",
    },
    {
        inputs: [],
        name: "beacon",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
        constant: true,
    },
    {
        inputs: [],
        name: "owner",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
        constant: true,
    },
    {
        inputs: [],
        name: "renounceOwnership",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "newOwner",
                type: "address",
            },
        ],
        name: "transferOwnership",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "_name",
                type: "string",
            },
            {
                internalType: "string",
                name: "_symbol",
                type: "string",
            },
            {
                internalType: "string",
                name: "baseURI",
                type: "string",
            },
            {
                internalType: "string",
                name: "contractURI",
                type: "string",
            },
            {
                internalType: "uint256",
                name: "salt",
                type: "uint256",
            },
        ],
        name: "createToken",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "_name",
                type: "string",
            },
            {
                internalType: "string",
                name: "_symbol",
                type: "string",
            },
            {
                internalType: "string",
                name: "baseURI",
                type: "string",
            },
            {
                internalType: "string",
                name: "contractURI",
                type: "string",
            },
            {
                internalType: "uint256",
                name: "_salt",
                type: "uint256",
            },
        ],
        name: "getAddress",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
        constant: true,
    },
];
