// @rarible/sdk/build/sdk-blockchains/ethereum/token-id.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumTokenId = void 0;
const tslib_1 = require("tslib");
const types_1 = require("@rarible/types");
class EthereumTokenId {
    constructor(sdk, apis) {
        this.sdk = sdk;
        this.apis = apis;
        this.generateTokenId = this.generateTokenId.bind(this);
    }
    generateTokenId(_a) {
        return tslib_1.__awaiter(this, arguments, void 0, function* ({ collection, minter }) {
            const nftTokenId = yield this.apis.collection.generateTokenId({
                collection: collection,
                minter: minter,
            });
            return {
                tokenId: nftTokenId.tokenId.toString(),
                signature: nftTokenId["@type"] === "ETHEREUM"
                    ? {
                        v: nftTokenId.signature.v,
                        r: (0, types_1.toBinary)(nftTokenId.signature.r),
                        s: (0, types_1.toBinary)(nftTokenId.signature.s),
                    }
                    : {
                        v: 0,
                        r: (0, types_1.toBinary)("0x"),
                        s: (0, types_1.toBinary)("0x"),
                    },
            };
        });
    }
}
exports.EthereumTokenId = EthereumTokenId;
