// @rarible/sdk/build/sdk-blockchains/ethereum/transfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumTransfer = void 0;
const tslib_1 = require("tslib");
const action_1 = require("@rarible/action");
const types_1 = require("@rarible/types");
const sdk_transaction_1 = require("@rarible/sdk-transaction");
const common_1 = require("./common");
class EthereumTransfer {
    constructor(sdk, wallet, apis) {
        this.sdk = sdk;
        this.wallet = wallet;
        this.apis = apis;
        this.transfer = this.transfer.bind(this);
        this.transferBasic = this.transferBasic.bind(this);
    }
    transfer(prepare) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { contract, tokenId, domain } = (0, common_1.getEthereumItemId)(prepare.itemId);
            if (!(0, common_1.isEVMBlockchain)(domain)) {
                throw new Error(`Not an ethereum item: ${prepare.itemId}`);
            }
            const [item, collection] = yield Promise.all([
                this.apis.item.getItemById({ itemId: prepare.itemId }),
                this.apis.collection.getCollectionById({ collection: `${domain}:${contract}` }),
            ]);
            return {
                multiple: collection.type === "ERC1155",
                maxAmount: item.supply,
                nftData: {
                    nftCollection: item.collection ? (0, types_1.toUnionContractAddress)(item.collection) : undefined,
                },
                submit: action_1.Action.create({
                    id: "transfer",
                    run: (request) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                        yield (0, common_1.checkWalletBlockchain)(this.wallet, domain);
                        const amount = request.amount !== undefined ? (0, types_1.toBigNumber)(request.amount.toFixed()) : undefined;
                        const tx = yield this.sdk.nft.transfer({
                            contract: (0, types_1.toEVMAddress)(contract),
                            tokenId: tokenId,
                        }, (0, common_1.convertToEthereumAddress)(request.to), amount);
                        return new sdk_transaction_1.BlockchainEthereumTransaction(tx, yield (0, common_1.getWalletNetwork)(this.wallet));
                    }),
                }),
            };
        });
    }
    transferBasic(request) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.transfer(request);
            return response.submit(request);
        });
    }
}
exports.EthereumTransfer = EthereumTransfer;
