// @rarible/protocol-ethereum-sdk/build/order/fill-order/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExchangeWrapperOrderType = void 0;
var ExchangeWrapperOrderType;
(function (ExchangeWrapperOrderType) {
    ExchangeWrapperOrderType[ExchangeWrapperOrderType["RARIBLE_V2"] = 0] = "RARIBLE_V2";
    ExchangeWrapperOrderType[ExchangeWrapperOrderType["OPENSEA_V1"] = 1] = "OPENSEA_V1";
    ExchangeWrapperOrderType[ExchangeWrapperOrderType["SEAPORT_ADVANCED_ORDERS"] = 2] = "SEAPORT_ADVANCED_ORDERS";
    ExchangeWrapperOrderType[ExchangeWrapperOrderType["X2Y2"] = 3] = "X2Y2";
    ExchangeWrapperOrderType[ExchangeWrapperOrderType["LOOKSRARE_ORDERS"] = 4] = "LOOKSRARE_ORDERS";
    ExchangeWrapperOrderType[ExchangeWrapperOrderType["AAM"] = 5] = "AAM";
    ExchangeWrapperOrderType[ExchangeWrapperOrderType["SEAPORT_V14"] = 6] = "SEAPORT_V14";
    ExchangeWrapperOrderType[ExchangeWrapperOrderType["LOOKSRARE_V2_ORDERS"] = 7] = "LOOKSRARE_V2_ORDERS";
    ExchangeWrapperOrderType[ExchangeWrapperOrderType["SEAPORT_V15"] = 9] = "SEAPORT_V15";
    ExchangeWrapperOrderType[ExchangeWrapperOrderType["SEAPORT_V16"] = 10] = "SEAPORT_V16";
})(ExchangeWrapperOrderType || (exports.ExchangeWrapperOrderType = ExchangeWrapperOrderType = {}));
