// @rarible/aptos-wallet/esm/wallets/generic.js
import { __awaiter } from "tslib";
import { normalizeAptosAddress } from "@rarible/sdk-common";
import { Network } from "../domain";
export class AptosGenericSdkWallet {
    constructor(aptos, account) {
        this.aptos = aptos;
        this.account = account;
    }
    signMessage(msg) {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                //@todo test aptos wallet doesn't provide fullMessage and nonce
                message: msg,
                signature: this.account.sign(msg).toString(),
            };
        });
    }
    getAccountInfo() {
        return __awaiter(this, void 0, void 0, function* () {
            return {
                address: normalizeAptosAddress(this.account.accountAddress.toString()),
                publicKey: this.account.publicKey.toString(),
                //@todo "Account" doesn't include network field
                network: Network.Testnet,
            };
        });
    }
    signAndSubmitTransaction(payload) {
        return __awaiter(this, void 0, void 0, function* () {
            const tx = yield this.aptos.transaction.build.simple({
                sender: this.account.accountAddress,
                data: {
                    function: payload.function,
                    typeArguments: payload.typeArguments,
                    functionArguments: payload.arguments,
                },
            });
            const { hash } = yield this.aptos.signAndSubmitTransaction({
                signer: this.account,
                transaction: tx,
            });
            return { hash };
        });
    }
}
