// @rarible/sdk-transaction/build/flow/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockchainFlowTransaction = void 0;
const tslib_1 = require("tslib");
const api_client_1 = require("@rarible/api-client");
class BlockchainFlowTransaction {
    constructor(transaction, network) {
        this.transaction = transaction;
        this.network = network;
        this.blockchain = api_client_1.Blockchain.FLOW;
    }
    hash() {
        return this.transaction.txId;
    }
    wait() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return {
                blockchain: this.blockchain,
                hash: this.transaction.txId,
            };
        });
    }
    getTxLink() {
        switch (this.network) {
            case "mainnet":
                return `https://flowscan.org/transaction/${this.hash()}`;
            case "testnet":
                return `https://testnet.flowscan.org/transaction/${this.hash()}`;
            case "emulator":
                return "";
            default:
                throw new Error("Unsupported transaction network");
        }
    }
    get isEmpty() {
        return false;
    }
}
exports.BlockchainFlowTransaction = BlockchainFlowTransaction;
