// @rarible/tezos-sdk/dist/sales/sell.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bundle_sell_arg_v2 = exports.sell_arg_v2 = exports.sell_v2_batch = exports.sellBundle = exports.sellV2 = void 0;
var tezos_common_1 = require("@rarible/tezos-common");
var tezos_common_2 = require("@rarible/tezos-common");
function process_order(config, order) {
    return __awaiter(this, void 0, void 0, function () {
        var ft_type_res;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!(order.s_sale_asset_contract != undefined)) return [3 /*break*/, 2];
                    return [4 /*yield*/, (0, tezos_common_1.get_ft_type)(config, order.s_sale_asset_contract)];
                case 1:
                    ft_type_res = _a.sent();
                    if (ft_type_res != undefined) {
                        order.s_sale_type = ft_type_res;
                        order.s_sale_asset_token_id = (0, tezos_common_1.process_token_id)(order.s_sale_type, order.s_sale_asset_token_id);
                    }
                    _a.label = 2;
                case 2: return [2 /*return*/, order];
            }
        });
    });
}
function sellV2(provider, order) {
    return __awaiter(this, void 0, void 0, function () {
        var args, seller, processed_amount, approve_a, op, order_id;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    args = [];
                    return [4 /*yield*/, (0, tezos_common_2.get_address)(provider)];
                case 1:
                    seller = _a.sent();
                    return [4 /*yield*/, process_order(provider.config, order)];
                case 2:
                    order = _a.sent();
                    return [4 /*yield*/, (0, tezos_common_1.absolute_amount)(provider.config, order.s_sale.sale_amount, order.s_sale_type, order.s_sale_asset_contract, order.s_sale_asset_token_id)];
                case 3:
                    processed_amount = _a.sent();
                    return [4 /*yield*/, (0, tezos_common_1.approve_v2)(provider, seller, tezos_common_1.AssetTypeV2.FA2, provider.config.transfer_manager, order.s_asset_contract, order.s_asset_token_id)];
                case 4:
                    approve_a = _a.sent();
                    if (approve_a)
                        args = args.concat(approve_a);
                    args = args.concat(sell_arg_v2(provider, order, processed_amount));
                    if (args.length === 0) {
                        throw new Error("Empty array of sell args");
                    }
                    return [4 /*yield*/, (0, tezos_common_1.send_batch)(provider, args)];
                case 5:
                    op = _a.sent();
                    return [4 /*yield*/, op.confirmation()];
                case 6:
                    _a.sent();
                    return [4 /*yield*/, (0, tezos_common_1.await_order)(provider.config, "".concat(order.s_asset_contract, ":").concat(order.s_asset_token_id), op.hash, tezos_common_1.ProtocolActivity.LIST, seller, 20, 2000)];
                case 7:
                    order_id = _a.sent();
                    if (order_id == undefined) {
                        throw new Error("Order was not found");
                    }
                    return [2 /*return*/, order_id];
            }
        });
    });
}
exports.sellV2 = sellV2;
function sellBundle(provider, order) {
    return __awaiter(this, void 0, void 0, function () {
        var args, seller, processed_amount, _a, _b, bundleItem, approve_a, e_1_1, op;
        var e_1, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    args = [];
                    return [4 /*yield*/, (0, tezos_common_2.get_address)(provider)];
                case 1:
                    seller = _d.sent();
                    return [4 /*yield*/, (0, tezos_common_1.absolute_amount)(provider.config, order.s_sale.sale_amount, order.s_sale_type, order.s_sale_asset_contract, order.s_sale_asset_token_id)];
                case 2:
                    processed_amount = _d.sent();
                    _d.label = 3;
                case 3:
                    _d.trys.push([3, 8, 9, 10]);
                    _a = __values(order.bundle), _b = _a.next();
                    _d.label = 4;
                case 4:
                    if (!!_b.done) return [3 /*break*/, 7];
                    bundleItem = _b.value;
                    return [4 /*yield*/, (0, tezos_common_1.approve_v2)(provider, seller, tezos_common_1.AssetTypeV2.FA2, provider.config.transfer_manager, bundleItem.asset_contract, bundleItem.asset_token_id)];
                case 5:
                    approve_a = _d.sent();
                    if (approve_a)
                        args = args.concat(approve_a);
                    _d.label = 6;
                case 6:
                    _b = _a.next();
                    return [3 /*break*/, 4];
                case 7: return [3 /*break*/, 10];
                case 8:
                    e_1_1 = _d.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 10];
                case 9:
                    try {
                        if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                    }
                    finally { if (e_1) throw e_1.error; }
                    return [7 /*endfinally*/];
                case 10:
                    args = args.concat(bundle_sell_arg_v2(provider, order, processed_amount));
                    if (!(args.length != 0)) return [3 /*break*/, 13];
                    return [4 /*yield*/, (0, tezos_common_1.send_batch)(provider, args)];
                case 11:
                    op = _d.sent();
                    return [4 /*yield*/, op.confirmation()];
                case 12:
                    _d.sent();
                    return [2 /*return*/, op];
                case 13: return [2 /*return*/];
            }
        });
    });
}
exports.sellBundle = sellBundle;
function sell_v2_batch(provider, order_form) {
    return __awaiter(this, void 0, void 0, function () {
        var args, seller, order_form_1, order_form_1_1, order, processed_amount, approve_a, e_2_1, op;
        var e_2, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    args = [];
                    return [4 /*yield*/, (0, tezos_common_2.get_address)(provider)];
                case 1:
                    seller = _b.sent();
                    _b.label = 2;
                case 2:
                    _b.trys.push([2, 8, 9, 10]);
                    order_form_1 = __values(order_form), order_form_1_1 = order_form_1.next();
                    _b.label = 3;
                case 3:
                    if (!!order_form_1_1.done) return [3 /*break*/, 7];
                    order = order_form_1_1.value;
                    return [4 /*yield*/, (0, tezos_common_1.absolute_amount)(provider.config, order.s_sale.sale_amount, order.s_sale_type, order.s_sale_asset_contract, order.s_sale_asset_token_id)];
                case 4:
                    processed_amount = _b.sent();
                    return [4 /*yield*/, (0, tezos_common_1.approve_v2)(provider, seller, tezos_common_1.AssetTypeV2.FA2, provider.config.transfer_manager, order.s_asset_contract, order.s_asset_token_id)];
                case 5:
                    approve_a = _b.sent();
                    if (approve_a)
                        args = args.concat(approve_a);
                    args = args.concat(sell_arg_v2(provider, order, processed_amount));
                    _b.label = 6;
                case 6:
                    order_form_1_1 = order_form_1.next();
                    return [3 /*break*/, 3];
                case 7: return [3 /*break*/, 10];
                case 8:
                    e_2_1 = _b.sent();
                    e_2 = { error: e_2_1 };
                    return [3 /*break*/, 10];
                case 9:
                    try {
                        if (order_form_1_1 && !order_form_1_1.done && (_a = order_form_1.return)) _a.call(order_form_1);
                    }
                    finally { if (e_2) throw e_2.error; }
                    return [7 /*endfinally*/];
                case 10:
                    if (!(args.length != 0)) return [3 /*break*/, 13];
                    return [4 /*yield*/, (0, tezos_common_1.send_batch)(provider, args)];
                case 11:
                    op = _b.sent();
                    return [4 /*yield*/, op.confirmation()];
                case 12:
                    _b.sent();
                    return [2 /*return*/, op];
                case 13: return [2 /*return*/];
            }
        });
    });
}
exports.sell_v2_batch = sell_v2_batch;
function delay(ms) {
    return new Promise(function (resolve) { return setTimeout(resolve, ms); });
}
function sell_arg_v2(provider, order, processed_amount) {
    var parameter = {
        prim: "Pair",
        args: [
            {
                string: order.s_asset_contract,
            },
            {
                prim: "Pair",
                args: [
                    {
                        int: "".concat(order.s_asset_token_id),
                    },
                    {
                        prim: "Pair",
                        args: [
                            {
                                int: "".concat(order.s_sale_type),
                            },
                            {
                                prim: "Pair",
                                args: [
                                    {
                                        bytes: (0, tezos_common_1.getAsset)(order.s_sale_type, order.s_sale_asset_contract, order.s_sale_asset_token_id)
                                    },
                                    {
                                        prim: "Pair",
                                        args: [
                                            (0, tezos_common_1.parts_to_micheline)(order.s_sale.sale_origin_fees),
                                            {
                                                prim: "Pair",
                                                args: [
                                                    (0, tezos_common_1.parts_to_micheline)(order.s_sale.sale_payouts),
                                                    {
                                                        prim: "Pair",
                                                        args: [
                                                            {
                                                                int: "".concat(processed_amount),
                                                            },
                                                            {
                                                                prim: "Pair",
                                                                args: [
                                                                    {
                                                                        int: "".concat(order.s_sale.sale_asset_qty),
                                                                    },
                                                                    {
                                                                        prim: "Pair",
                                                                        args: [
                                                                            (0, tezos_common_1.optional_date_arg)(order.s_sale.sale_start),
                                                                            {
                                                                                prim: "Pair",
                                                                                args: [
                                                                                    (0, tezos_common_1.optional_date_arg)(order.s_sale.sale_end),
                                                                                    {
                                                                                        prim: "Pair",
                                                                                        args: [
                                                                                            {
                                                                                                int: "".concat(order.s_sale.sale_max_fees_base_boint),
                                                                                            },
                                                                                            {
                                                                                                prim: "Pair",
                                                                                                args: [
                                                                                                    {
                                                                                                        prim: "None",
                                                                                                    },
                                                                                                    {
                                                                                                        prim: "None",
                                                                                                    },
                                                                                                ],
                                                                                            },
                                                                                        ],
                                                                                    },
                                                                                ],
                                                                            },
                                                                        ],
                                                                    },
                                                                ],
                                                            },
                                                        ],
                                                    },
                                                ],
                                            },
                                        ],
                                    },
                                ],
                            },
                        ],
                    },
                ],
            },
        ],
    };
    return { destination: provider.config.sales, entrypoint: "sell", parameter: parameter };
}
exports.sell_arg_v2 = sell_arg_v2;
function bundle_sell_arg_v2(provider, order, processed_amount) {
    var parameter = {
        prim: "Pair",
        args: [
            {
                bytes: (0, tezos_common_1.mkPackedBundle)(order.bundle)
            },
            {
                prim: "Pair",
                args: [
                    {
                        int: "".concat(order.s_sale_type)
                    },
                    {
                        prim: "Pair",
                        args: [
                            {
                                bytes: "".concat((0, tezos_common_1.getAsset)(order.s_sale_type, order.s_sale_asset_contract, order.s_sale_asset_token_id))
                            },
                            {
                                prim: "Pair",
                                args: [
                                    (0, tezos_common_1.parts_to_micheline)(order.s_sale.sale_origin_fees),
                                    {
                                        prim: "Pair",
                                        args: [
                                            (0, tezos_common_1.parts_to_micheline)(order.s_sale.sale_payouts),
                                            {
                                                prim: "Pair",
                                                args: [
                                                    {
                                                        int: "".concat(processed_amount)
                                                    },
                                                    {
                                                        prim: "Pair",
                                                        args: [
                                                            (0, tezos_common_1.optional_date_arg)(order.s_sale.sale_start),
                                                            {
                                                                prim: "Pair",
                                                                args: [
                                                                    (0, tezos_common_1.optional_date_arg)(order.s_sale.sale_end),
                                                                    {
                                                                        prim: "Pair",
                                                                        args: [
                                                                            {
                                                                                int: "".concat(order.s_sale.sale_qty)
                                                                            },
                                                                            {
                                                                                prim: "Pair",
                                                                                args: [
                                                                                    {
                                                                                        int: "".concat(order.s_sale.sale_max_fees_base_boint)
                                                                                    },
                                                                                    {
                                                                                        prim: "Pair",
                                                                                        args: [
                                                                                            {
                                                                                                prim: "None"
                                                                                            },
                                                                                            {
                                                                                                prim: "None"
                                                                                            }
                                                                                        ]
                                                                                    }
                                                                                ]
                                                                            }
                                                                        ]
                                                                    }
                                                                ]
                                                            }
                                                        ]
                                                    }
                                                ]
                                            }
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
                ]
            }
        ]
    };
    return { destination: provider.config.sales, entrypoint: "sell_bundle", parameter: parameter };
}
exports.bundle_sell_arg_v2 = bundle_sell_arg_v2;
