// @rarible/protocol-ethereum-sdk/build/nft/contracts/erc721/deploy/rarible-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.erc721RaribleFactoryABI = exports.createTestRaribleFactoryContract = exports.createErc721FactoryContract = void 0;
const contracts_1 = require("../../../../common/contracts");
function createErc721FactoryContract(ethereum, address) {
    return ethereum.createContract(exports.erc721RaribleFactoryABI, address);
}
exports.createErc721FactoryContract = createErc721FactoryContract;
function createTestRaribleFactoryContract(web3, address) {
    return new web3.eth.Contract(exports.erc721RaribleFactoryABI, address, contracts_1.NumberDataFormat);
}
exports.createTestRaribleFactoryContract = createTestRaribleFactoryContract;
exports.erc721RaribleFactoryABI = [
    {
        inputs: [
            {
                internalType: "address",
                name: "_beacon",
                type: "address",
            },
            {
                internalType: "address",
                name: "_transferProxy",
                type: "address",
            },
            {
                internalType: "address",
                name: "_lazyTransferProxy",
                type: "address",
            },
        ],
        stateMutability: "nonpayable",
        type: "constructor",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "proxy",
                type: "address",
            },
        ],
        name: "Create721RaribleProxy",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "previousOwner",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "newOwner",
                type: "address",
            },
        ],
        name: "OwnershipTransferred",
        type: "event",
    },
    {
        inputs: [],
        name: "beacon",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
        constant: true,
    },
    {
        inputs: [],
        name: "owner",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
        constant: true,
    },
    {
        inputs: [],
        name: "renounceOwnership",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "newOwner",
                type: "address",
            },
        ],
        name: "transferOwnership",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "_name",
                type: "string",
            },
            {
                internalType: "string",
                name: "_symbol",
                type: "string",
            },
            {
                internalType: "string",
                name: "baseURI",
                type: "string",
            },
            {
                internalType: "string",
                name: "contractURI",
                type: "string",
            },
            {
                internalType: "uint256",
                name: "salt",
                type: "uint256",
            },
        ],
        name: "createToken",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "_name",
                type: "string",
            },
            {
                internalType: "string",
                name: "_symbol",
                type: "string",
            },
            {
                internalType: "string",
                name: "baseURI",
                type: "string",
            },
            {
                internalType: "string",
                name: "contractURI",
                type: "string",
            },
            {
                internalType: "uint256",
                name: "_salt",
                type: "uint256",
            },
        ],
        name: "getAddress",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
        constant: true,
    },
];
