// @rarible/api-client/build/models/Blockchain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Blockchain = void 0;
var Blockchain;
(function (Blockchain) {
    Blockchain["APTOS"] = "APTOS";
    Blockchain["ETHEREUM"] = "ETHEREUM";
    Blockchain["POLYGON"] = "POLYGON";
    Blockchain["FLOW"] = "FLOW";
    Blockchain["TEZOS"] = "TEZOS";
    Blockchain["SOLANA"] = "SOLANA";
    Blockchain["IMMUTABLEX"] = "IMMUTABLEX";
    Blockchain["MANTLE"] = "MANTLE";
    Blockchain["ARBITRUM"] = "ARBITRUM";
    Blockchain["CHILIZ"] = "CHILIZ";
    Blockchain["LIGHTLINK"] = "LIGHTLINK";
    Blockchain["ZKSYNC"] = "ZKSYNC";
    Blockchain["ASTARZKEVM"] = "ASTARZKEVM";
    Blockchain["BASE"] = "BASE";
    Blockchain["RARI"] = "RARI";
    Blockchain["CELO"] = "CELO";
    Blockchain["FIEF"] = "FIEF";
    Blockchain["XAI"] = "XAI";
    Blockchain["KROMA"] = "KROMA";
    Blockchain["ZKLINK"] = "ZKLINK";
    Blockchain["OASYS"] = "OASYS";
    Blockchain["QUAI"] = "QUAI";
    Blockchain["ECLIPSE"] = "ECLIPSE";
    Blockchain["SAAKURU"] = "SAAKURU";
    Blockchain["OASIS"] = "OASIS";
    Blockchain["PALM"] = "PALM";
    Blockchain["MATCH"] = "MATCH";
    Blockchain["FIVIRE"] = "FIVIRE";
    Blockchain["SEI"] = "SEI";
    Blockchain["CAMP"] = "CAMP";
    Blockchain["LISK"] = "LISK";
    Blockchain["MOONBEAM"] = "MOONBEAM";
    Blockchain["ETHERLINK"] = "ETHERLINK";
    Blockchain["ZKCANDY"] = "ZKCANDY";
    Blockchain["ALEPHZERO"] = "ALEPHZERO";
    Blockchain["BERACHAIN"] = "BERACHAIN";
    Blockchain["ABSTRACT"] = "ABSTRACT";
    Blockchain["SHAPE"] = "SHAPE";
    Blockchain["TELOS"] = "TELOS";
    Blockchain["HEDERAEVM"] = "HEDERAEVM";
    Blockchain["VICTION"] = "VICTION";
    Blockchain["SETTLUS"] = "SETTLUS";
    Blockchain["GOAT"] = "GOAT";
    Blockchain["HYPEREVM"] = "HYPEREVM";
    Blockchain["HEDERA"] = "HEDERA";
    Blockchain["CROSSFI"] = "CROSSFI";
})(Blockchain = exports.Blockchain || (exports.Blockchain = {}));
