// @rarible/api-client/build/models/Collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CollectionFeatures = exports.CollectionStatus = exports.CollectionType = exports.CollectionStructure = void 0;
var CollectionStructure;
(function (CollectionStructure) {
    CollectionStructure["REGULAR"] = "REGULAR";
    CollectionStructure["COMPOSITE"] = "COMPOSITE";
    CollectionStructure["PART"] = "PART";
})(CollectionStructure = exports.CollectionStructure || (exports.CollectionStructure = {}));
var CollectionType;
(function (CollectionType) {
    CollectionType["CRYPTO_PUNKS"] = "CRYPTO_PUNKS";
    CollectionType["ERC721"] = "ERC721";
    CollectionType["ERC1155"] = "ERC1155";
    CollectionType["FLOW"] = "FLOW";
    CollectionType["TEZOS_NFT"] = "TEZOS_NFT";
    CollectionType["TEZOS_MT"] = "TEZOS_MT";
    CollectionType["SOLANA"] = "SOLANA";
    CollectionType["TOKEN_GROUP_2022"] = "TOKEN_GROUP_2022";
    CollectionType["IMMUTABLEX"] = "IMMUTABLEX";
    CollectionType["APTOS"] = "APTOS";
    CollectionType["HEDERA"] = "HEDERA";
})(CollectionType = exports.CollectionType || (exports.CollectionType = {}));
var CollectionStatus;
(function (CollectionStatus) {
    CollectionStatus["PENDING"] = "PENDING";
    CollectionStatus["ERROR"] = "ERROR";
    CollectionStatus["CONFIRMED"] = "CONFIRMED";
})(CollectionStatus = exports.CollectionStatus || (exports.CollectionStatus = {}));
var CollectionFeatures;
(function (CollectionFeatures) {
    CollectionFeatures["APPROVE_FOR_ALL"] = "APPROVE_FOR_ALL";
    CollectionFeatures["SET_URI_PREFIX"] = "SET_URI_PREFIX";
    CollectionFeatures["BURN"] = "BURN";
    CollectionFeatures["MINT_WITH_ADDRESS"] = "MINT_WITH_ADDRESS";
    CollectionFeatures["SECONDARY_SALE_FEES"] = "SECONDARY_SALE_FEES";
    CollectionFeatures["MINT_AND_TRANSFER"] = "MINT_AND_TRANSFER";
    CollectionFeatures["PAUSABLE"] = "PAUSABLE";
    CollectionFeatures["NOT_FOR_SALE"] = "NOT_FOR_SALE";
})(CollectionFeatures = exports.CollectionFeatures || (exports.CollectionFeatures = {}));
