// @rarible/api-client/build/models/OlapBlockchain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OlapBlockchain = void 0;
var OlapBlockchain;
(function (OlapBlockchain) {
    OlapBlockchain["ETHEREUM"] = "ETHEREUM";
    OlapBlockchain["FLOW"] = "FLOW";
    OlapBlockchain["TEZOS"] = "TEZOS";
    OlapBlockchain["POLYGON"] = "POLYGON";
    OlapBlockchain["SOLANA"] = "SOLANA";
    OlapBlockchain["IMMUTABLEX"] = "IMMUTABLEX";
    OlapBlockchain["MANTLE"] = "MANTLE";
    OlapBlockchain["ARBITRUM"] = "ARBITRUM";
    OlapBlockchain["CHILIZ"] = "CHILIZ";
    OlapBlockchain["LIGHTLINK"] = "LIGHTLINK";
    OlapBlockchain["ZKSYNC"] = "ZKSYNC";
    OlapBlockchain["ASTARZKEVM"] = "ASTARZKEVM";
    OlapBlockchain["BASE"] = "BASE";
    OlapBlockchain["RARI"] = "RARI";
    OlapBlockchain["CELO"] = "CELO";
    OlapBlockchain["FIEF"] = "FIEF";
    OlapBlockchain["XAI"] = "XAI";
    OlapBlockchain["KROMA"] = "KROMA";
    OlapBlockchain["APTOS"] = "APTOS";
    OlapBlockchain["ZKLINK"] = "ZKLINK";
    OlapBlockchain["OASYS"] = "OASYS";
    OlapBlockchain["QUAI"] = "QUAI";
    OlapBlockchain["ECLIPSE"] = "ECLIPSE";
    OlapBlockchain["SAAKURU"] = "SAAKURU";
    OlapBlockchain["OASIS"] = "OASIS";
    OlapBlockchain["PALM"] = "PALM";
    OlapBlockchain["MATCH"] = "MATCH";
    OlapBlockchain["FIVIRE"] = "FIVIRE";
    OlapBlockchain["SEI"] = "SEI";
    OlapBlockchain["CAMP"] = "CAMP";
    OlapBlockchain["LISK"] = "LISK";
    OlapBlockchain["MOONBEAM"] = "MOONBEAM";
    OlapBlockchain["ETHERLINK"] = "ETHERLINK";
    OlapBlockchain["ZKCANDY"] = "ZKCANDY";
    OlapBlockchain["ALEPHZERO"] = "ALEPHZERO";
    OlapBlockchain["BERACHAIN"] = "BERACHAIN";
    OlapBlockchain["ABSTRACT"] = "ABSTRACT";
    OlapBlockchain["SHAPE"] = "SHAPE";
    OlapBlockchain["TELOS"] = "TELOS";
    OlapBlockchain["HEDERAEVM"] = "HEDERAEVM";
    OlapBlockchain["VICTION"] = "VICTION";
    OlapBlockchain["SETTLUS"] = "SETTLUS";
    OlapBlockchain["GOAT"] = "GOAT";
    OlapBlockchain["HYPEREVM"] = "HYPEREVM";
    OlapBlockchain["HEDERA"] = "HEDERA";
    OlapBlockchain["CROSSFI"] = "CROSSFI";
})(OlapBlockchain = exports.OlapBlockchain || (exports.OlapBlockchain = {}));
