// @rarible/api-client/build/models/OlapCurrencyType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OlapCurrencyType = void 0;
var OlapCurrencyType;
(function (OlapCurrencyType) {
    OlapCurrencyType["USD"] = "USD";
    OlapCurrencyType["ETH"] = "ETH";
    OlapCurrencyType["XTZ"] = "XTZ";
    OlapCurrencyType["FLOW"] = "FLOW";
    OlapCurrencyType["MATIC"] = "MATIC";
    OlapCurrencyType["POL"] = "POL";
    OlapCurrencyType["SOL"] = "SOL";
    OlapCurrencyType["MNT"] = "MNT";
    OlapCurrencyType["CHZ"] = "CHZ";
    OlapCurrencyType["CELO"] = "CELO";
    OlapCurrencyType["FIEF"] = "FIEF";
    OlapCurrencyType["XAI"] = "XAI";
    OlapCurrencyType["APT"] = "APT";
    OlapCurrencyType["OAS"] = "OAS";
    OlapCurrencyType["QUAI"] = "QUAI";
    OlapCurrencyType["ROSE"] = "ROSE";
    OlapCurrencyType["BNB"] = "BNB";
    OlapCurrencyType["PALM"] = "PALM";
    OlapCurrencyType["FIVIRE"] = "FIVIRE";
    OlapCurrencyType["SEI"] = "SEI";
    OlapCurrencyType["GLMR"] = "GLMR";
    OlapCurrencyType["AZERO"] = "AZERO";
    OlapCurrencyType["BERA"] = "BERA";
    OlapCurrencyType["TLOS"] = "TLOS";
    OlapCurrencyType["HBAR"] = "HBAR";
    OlapCurrencyType["VIC"] = "VIC";
    OlapCurrencyType["BTC"] = "BTC";
    OlapCurrencyType["HYPE"] = "HYPE";
    OlapCurrencyType["XFI"] = "XFI";
})(OlapCurrencyType = exports.OlapCurrencyType || (exports.OlapCurrencyType = {}));
