// @rarible/aptos-wallet/build/wallets/external.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AptosSdkWallet = void 0;
const tslib_1 = require("tslib");
const types_1 = require("@rarible/types");
const sdk_common_1 = require("@rarible/sdk-common");
const common_1 = require("./common");
class AptosSdkWallet {
    constructor(account) {
        this.account = account;
    }
    signMessage(msg, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { signature, fullMessage } = yield this.account.signMessage({
                message: msg,
                nonce: (options === null || options === void 0 ? void 0 : options.nonce) || (0, types_1.randomWord)(),
            });
            if (Array.isArray(signature)) {
                return {
                    signature: signature[0].toString(),
                    message: fullMessage,
                };
            }
            return {
                signature: signature.toString(),
                message: fullMessage.toString(),
            };
        });
    }
    getAccountInfo() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.account.account)
                throw new Error("Cannot get account info: account has not been set");
            const { address, publicKey } = yield this.account.account();
            return {
                address: (0, sdk_common_1.normalizeAptosAddress)(address),
                publicKey: Array.isArray(publicKey) ? publicKey[0] : publicKey.toString(),
                network: (0, common_1.normalizeAptosNetwork)(yield this.account.network()),
            };
        });
    }
    getPublicKey() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const account = yield this.getAccountInfo();
            return account.publicKey;
        });
    }
    signAndSubmitTransaction(payload) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { hash } = yield this.account.signAndSubmitTransaction({
                arguments: payload.arguments,
                function: payload.function,
                type_arguments: payload.typeArguments,
            });
            return { hash };
        });
    }
}
exports.AptosSdkWallet = AptosSdkWallet;
