// @rarible/aptos-wallet/build/wallets/generic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AptosGenericSdkWallet = void 0;
const tslib_1 = require("tslib");
const sdk_common_1 = require("@rarible/sdk-common");
const domain_1 = require("../domain");
class AptosGenericSdkWallet {
    constructor(aptos, account) {
        this.aptos = aptos;
        this.account = account;
    }
    signMessage(msg) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return {
                //@todo test aptos wallet doesn't provide fullMessage and nonce
                message: msg,
                signature: this.account.sign(msg).toString(),
            };
        });
    }
    getAccountInfo() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return {
                address: (0, sdk_common_1.normalizeAptosAddress)(this.account.accountAddress.toString()),
                publicKey: this.account.publicKey.toString(),
                //@todo "Account" doesn't include network field
                network: domain_1.Network.Testnet,
            };
        });
    }
    signAndSubmitTransaction(payload) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const tx = yield this.aptos.transaction.build.simple({
                sender: this.account.accountAddress,
                data: {
                    function: payload.function,
                    typeArguments: payload.typeArguments,
                    functionArguments: payload.arguments,
                },
            });
            const { hash } = yield this.aptos.signAndSubmitTransaction({
                signer: this.account,
                transaction: tx,
            });
            return { hash };
        });
    }
}
exports.AptosGenericSdkWallet = AptosGenericSdkWallet;
