// @rarible/aptos-wallet/build/wallets/wallet-core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AptosWalletCore = void 0;
const tslib_1 = require("tslib");
const types_1 = require("@rarible/types");
const sdk_common_1 = require("@rarible/sdk-common");
const common_1 = require("./common");
class AptosWalletCore {
    constructor(wallet) {
        this.wallet = wallet;
    }
    signMessage(message, options) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.wallet.signMessage({
                message,
                nonce: (options === null || options === void 0 ? void 0 : options.nonce) || (0, types_1.randomWord)(),
            });
            if (Array.isArray(response.signature)) {
                return {
                    signature: response.signature[0].toString(),
                    message: response.fullMessage.toString(),
                };
            }
            return {
                signature: response.signature.toString(),
                message: response.fullMessage.toString(),
            };
        });
    }
    getAccountInfo() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const accountInfo = this.wallet.account;
            if (!accountInfo || !this.wallet.network) {
                throw new Error("AccountInfo does not exist");
            }
            return {
                address: (0, sdk_common_1.normalizeAptosAddress)(accountInfo.address),
                publicKey: Array.isArray(accountInfo.publicKey) ? accountInfo.publicKey[0] : accountInfo.publicKey,
                network: (0, common_1.normalizeAptosNetwork)(this.wallet.network.name),
            };
        });
    }
    signAndSubmitTransaction(payload) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const { hash } = yield this.wallet.signAndSubmitTransaction({
                data: {
                    functionArguments: payload.arguments,
                    function: payload.function,
                    typeArguments: payload.typeArguments,
                },
            });
            return { hash };
        });
    }
}
exports.AptosWalletCore = AptosWalletCore;
