// @rarible/sdk-common/build/logger/overrides.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBlockchainByConnectorId = exports.isFlowWarning = exports.FLOW_WARN_MESSAGES = exports.isSolanaWarning = exports.isTezosWarning = exports.isEVMWarning = exports.isInfoLevel = exports.isCancelMessage = exports.isCancelCode = exports.OUT_OF_GAS_ERROR = exports.COMMON_INFO_MESSAGES = exports.CANCEL_MESSAGES = exports.INVALID_TX_PARAMS_EIP_1559_ERROR = void 0;
const api_client_1 = require("@rarible/api-client");
exports.INVALID_TX_PARAMS_EIP_1559_ERROR = "Invalid transaction params: params specify an EIP-1559 transaction but the current network does not support EIP-1559";
const EVM_WARN_MESSAGES = [
    exports.INVALID_TX_PARAMS_EIP_1559_ERROR,
    "underlying network changed",
    "Balance not enough to cover gas fee. Please deposit at least",
    "Biaya gas telah diperbarui dan Anda memerlukan",
    "err: max fee per gas less than block base fee",
    "Error while gas estimation with message cannot estimate gas",
    "transaction may fail may require manual",
    "gas limit",
    "gas required exceeds allowance",
    "Insufficient fee balance",
    "Insufficient ETH to pay the network fees",
    "insufficient funds for gas * price + value",
    "max priority fee per gas higher than max fee per gas",
    "insufficient funds for intrinsic transaction cost",
    "intrinsic gas too low",
    "max fee per gas less than block base fee",
    "maxFeePerGas cannot be less than maxPriorityFeePerGas",
    "Please deposit at least",
    "replacement transaction underpriced",
    "Returned error: insufficient funds for gas * price + value",
    "Returned error: transaction underpriced",
    "Saldo tidak cukup untuk menutup biaya gas. Harap setor setidaknya",
    "The gas price is low, please increase the gas price try again",
    "transaction underpriced",
    "Комиссия за газ обновлена, и вам необходимо",
    "Insufficient ETH funds",
    "Please enable Blind signing or Contract data in the Ethereum app Settings",
    "No keyring found for the requested account. Error info: There are keyrings, but none match the address",
    "Link Window Closed",
    "nonce too low",
    "transaction would cause overdraft",
    "PromiEvent timeout",
].map(msg => msg.toLowerCase());
exports.CANCEL_MESSAGES = [
    "Transaction canceled",
    "Request canceled by user",
    "User canceled",
    "Request cancelled by user",
    "Cancelled by User",
    "Request has been cancelled by the user",
    "transaction was cancelled",
    "transaction was canceled",
    "Sign transaction cancelled",
    "Firma de transacción cancelada",
    "Sign transaction cancelled",
    "Signing transaction was cancelled",
    "Transação de assinatura cancelada",
    "You canceled",
    "User cancelled login",
    "User cancelled the request",
    "user reject this request",
    "MetaMask Tx Signature: User refused to sign the transaction.",
    "Request rejected",
    "User rejected the transaction",
    "Please enable Blind Signature or Contract Data in Ethereum Application Settings.",
    "User denied to sign transaction",
    "User declined the request.",
    "PocketUniverse Tx Signature: User declined to sign the transaction.",
    "Transaction declined",
    "User declined transaction",
    "MetaMask Tx Signature: Transaction signature denied by user.",
    "Transaction rejected",
    "User refused to sign the transaction",
    "Please enable Blind Signing or Contract Data in Ethereum Application Settings",
    "user rejected the request",
    "PocketUniverse Tx Signature: The user rejected the transaction signature.",
    "iFrame link is closed",
    "Link iFrame Closed",
    "rejected request from DeFi Wallet",
    "User rejected methods",
    "Транзакция отменена",
    "Failed to sign transaction",
    "Permission not given for signing message",
    "User denied message signature",
    "User denied transaction signature",
    "Подписание транзакции отменено",
    "การลงนามธุรกรรมถูกยกเลิก",
    "ยกเลิกแล้ว",
    "用户取消了操作",
    "签署交易已取消",
    "授权被拒绝",
    "membatalkan",
    "La transaction de signature a été annulée",
    "User denied account authorization",
    "Firma transazione annullata",
    "İşlem imzalama iptal edildi",
    "Nutzer hat die Transaktion abgelehnt",
    "Signature de transaction annulée",
    "Signiervorgang abgebrochen",
    "Reject by the user",
    "User closed modal",
    "Permission denied",
    "Approval Denied",
    "The requested account and/or method has not been authorized by the user",
    "user did not approve",
    "Der Nutzer hat die Anfrage abgelehnt",
    "The user rejected the request",
    "El usuario rechazó la solicitud",
    "L'utilisateur a rejeté la demande",
    "O usuário rejeitou a solicitação",
    "The user rejected the request through Exodus",
    "Permission denied, denied",
    "user closed popup",
    "User has rejected the request",
    "The user rejected the request",
].map(msg => msg.toLowerCase());
exports.COMMON_INFO_MESSAGES = [
    "The gas fee has been updated and you need",
    "The action was aborted by the user",
    "User denied requested chains",
    "Popup closed",
    "Please verify email address",
    "User denied account access",
    "Connection request reset. Please try again",
    "The tab is not active",
    "An internal error has occurred",
    "Произошла внутренняя ошибка.",
    "内部エラーが発生しました",
    "Une erreur interne s'est produite",
    "Tab inaktiv",
    "An internal error has occurred",
    "Ett internt fel har inträffat",
    "L'onglet n'est pas actif",
    "Het tabblad is niet actief",
    "Ha ocurrido un error interno",
    "A apărut o eroare internă",
    "Ha ocurrido un error interno",
    "La pestaña no está activa",
    "Si è verificato un errore interno",
    "Vyskytla sa interná chyba",
    "Вкладка не активна",
    "خطایی داخلی رخ داده است",
    "内部エラーが発生しました",
    "A guia não está ativa",
    "Dahili bir hata oluştu",
    "Er is een interne fout opgetreden",
    "Sekme etkin değil",
    "زبانه فعال نیست",
    "发生了一个内部的错误",
    "發生內部錯誤",
    "User Reject",
].map(msg => msg.toLowerCase());
exports.OUT_OF_GAS_ERROR = "returned values aren't valid, did it run out of gas";
const shortCancelMessages = ["cancel", "canceled", "cancelled", "rejected"];
function isCancelCode(code) {
    return code === 4001 || code === 4100 || code === "ACTION_REJECTED";
}
exports.isCancelCode = isCancelCode;
function isCancelMessage(msg, isLowerCase = false) {
    if (!msg || typeof msg !== "string") {
        return false;
    }
    let msgLowerCase = isLowerCase ? msg : msg.toLowerCase();
    if (shortCancelMessages.includes(msgLowerCase)) {
        return true;
    }
    return exports.CANCEL_MESSAGES.some(msg => msgLowerCase === null || msgLowerCase === void 0 ? void 0 : msgLowerCase.includes(msg));
}
exports.isCancelMessage = isCancelMessage;
function isInfoLevel(error) {
    var _a;
    if (!(error === null || error === void 0 ? void 0 : error.message) || typeof (error === null || error === void 0 ? void 0 : error.message) !== "string") {
        return false;
    }
    const msgLowerCase = error === null || error === void 0 ? void 0 : error.message.toLowerCase();
    if (isCancelCode((_a = error === null || error === void 0 ? void 0 : error.error) === null || _a === void 0 ? void 0 : _a.code) || isCancelMessage(msgLowerCase, true)) {
        return true;
    }
    return exports.COMMON_INFO_MESSAGES.some(msg => msgLowerCase === null || msgLowerCase === void 0 ? void 0 : msgLowerCase.includes(msg));
}
exports.isInfoLevel = isInfoLevel;
function isEVMWarning(error) {
    if ((error === null || error === void 0 ? void 0 : error.name) && ["WrongNetworkWarning", "InsufficientFundsError"].includes(error === null || error === void 0 ? void 0 : error.name))
        return true;
    const msgLowerCase = error === null || error === void 0 ? void 0 : error.message.toLowerCase();
    return EVM_WARN_MESSAGES.some(msg => msgLowerCase === null || msgLowerCase === void 0 ? void 0 : msgLowerCase.includes(msg));
}
exports.isEVMWarning = isEVMWarning;
function isTezosWarning(err) {
    var _a;
    const isWrappedError = err.name === "TezosProviderError";
    const originalError = isWrappedError ? err.error : err;
    return (((originalError === null || originalError === void 0 ? void 0 : originalError.name) === "UnknownBeaconError" && (originalError === null || originalError === void 0 ? void 0 : originalError.title) === "Aborted") ||
        (originalError === null || originalError === void 0 ? void 0 : originalError.name) === "NotGrantedTempleWalletError" ||
        (originalError === null || originalError === void 0 ? void 0 : originalError.name) === "NoAddressBeaconError" ||
        (originalError === null || originalError === void 0 ? void 0 : originalError.name) === "NoPrivateKeyBeaconError" ||
        (originalError === null || originalError === void 0 ? void 0 : originalError.name) === "BroadcastBeaconError" ||
        (originalError === null || originalError === void 0 ? void 0 : originalError.name) === "MissedBlockDuringConfirmationError" ||
        (originalError === null || originalError === void 0 ? void 0 : originalError.message) === "Error: timeout of 30000ms exceeded" ||
        ((_a = err === null || err === void 0 ? void 0 : err.message) === null || _a === void 0 ? void 0 : _a.endsWith("does not have enough funds for transaction")));
}
exports.isTezosWarning = isTezosWarning;
function isSolanaWarning(error) {
    var _a;
    return ((error === null || error === void 0 ? void 0 : error.name) === "User rejected the request." || ((_a = error === null || error === void 0 ? void 0 : error.error) === null || _a === void 0 ? void 0 : _a.code) === 4001 || (error === null || error === void 0 ? void 0 : error.message) === "Approval Denied");
}
exports.isSolanaWarning = isSolanaWarning;
exports.FLOW_WARN_MESSAGES = ["[Error Code: 1007] invalid proposal key", "User rejected signature"];
function isFlowWarning(error) {
    return exports.FLOW_WARN_MESSAGES.some(msg => { var _a; return (_a = error === null || error === void 0 ? void 0 : error.message) === null || _a === void 0 ? void 0 : _a.includes(msg); });
}
exports.isFlowWarning = isFlowWarning;
function getBlockchainByConnectorId(providerId) {
    switch (providerId) {
        case "beacon":
            return api_client_1.BlockchainGroup.TEZOS;
        case "fcl":
        case "mattel":
            return api_client_1.BlockchainGroup.FLOW;
        case "phantom":
        case "solflare":
            return api_client_1.BlockchainGroup.SOLANA;
        case "injected":
        case "fortmatic":
        case "iframe":
        case "immutablex":
        case "mew":
        case "portis":
        case "torus":
        case "firebase":
        case "firebase-apple":
        case "firebase-email":
        case "walletconnect":
        case "walletlink":
            return api_client_1.BlockchainGroup.ETHEREUM;
        default:
            return undefined;
    }
}
exports.getBlockchainByConnectorId = getBlockchainByConnectorId;
