// @rarible/immutable-sdk/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBalance = exports.IMX_ENV_CONFIG = exports.IMX_CONFIG = exports.createImxSdk = void 0;
const transfer_1 = require("./nft/transfer");
const order_1 = require("./order");
const burn_1 = require("./nft/burn");
const balance_1 = require("./balance/balance");
const apis_1 = require("./apis");
function createImxSdk(link, environment, config = {}) {
    const apis = (0, apis_1.createApis)(environment, config === null || config === void 0 ? void 0 : config.apiClientParams);
    return {
        nft: {
            transfer: transfer_1.transfer.bind(null, link),
            burn: burn_1.burn.bind(null, link),
        },
        order: {
            sell: order_1.sell.bind(null, link),
            buy: order_1.buy.bind(null, environment, link, apis),
            cancel: order_1.cancel.bind(null, link),
        },
        balance: {
            getBalance: balance_1.getBalance.bind(null, environment, apis),
        },
    };
}
exports.createImxSdk = createImxSdk;
var env_1 = require("./config/env");
Object.defineProperty(exports, "IMX_CONFIG", { enumerable: true, get: function () { return env_1.IMX_CONFIG; } });
Object.defineProperty(exports, "IMX_ENV_CONFIG", { enumerable: true, get: function () { return env_1.IMX_ENV_CONFIG; } });
var balance_2 = require("./balance/balance");
Object.defineProperty(exports, "getBalance", { enumerable: true, get: function () { return balance_2.getBalance; } });
