// @rarible/eclipse-sdk/build/sdk/sdk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EclipseSdk = void 0;
const web3_js_1 = require("@solana/web3.js");
const debug_logger_1 = require("../logger/debug-logger");
const balance_1 = require("./balance/balance");
const nft_1 = require("./nft/nft");
const account_1 = require("./account/account");
const order_1 = require("./order/order");
class EclipseSdk {
    constructor(connection, cluster, debug = false) {
        this.connection = connection;
        this.cluster = cluster;
        this.debug = debug;
        this.debugLogger = new debug_logger_1.DebugLogger(debug);
        this.balances = new balance_1.EclipseBalancesSdk(connection);
        this.account = new account_1.EclipseAccountSdk(connection, this.debugLogger);
        this.nft = new nft_1.EclipseNftSdk(connection, this.debugLogger, this.account);
        this.order = new order_1.EclipseOrderSdk(connection, this.debugLogger);
    }
    static create(config) {
        var _a, _b;
        const connection = new web3_js_1.Connection((_a = config.connection.endpoint) !== null && _a !== void 0 ? _a : (0, web3_js_1.clusterApiUrl)(config.connection.cluster), (_b = config.connection.commitmentOrConfig) !== null && _b !== void 0 ? _b : "confirmed");
        return new EclipseSdk(connection, config.connection.cluster, config.debug);
    }
    confirmTransaction(...args) {
        return this.connection.confirmTransaction(...args);
    }
}
exports.EclipseSdk = EclipseSdk;
