// @rarible/protocol-ethereum-sdk/build/config/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.baseConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.baseConfig = {
    basePath: "https://base-api.rarible.org",
    chainId: 8453,
    environment: "production",
    blockchain: api_client_1.Blockchain.BASE,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x6C65a3C3AA67b126e43F86DA85775E0F5e9743F7"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x6563a331A411829918025D8a7e1d348f8b250906"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x4217a346C8b48731641327b65bb6F6d3243d64e2"),
        erc20: (0, types_1.toEVMAddress)("0x13b05523634ABb96E6017Da71b7698CAecDf50b2"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x339e61eb644A29B134D7fD3fA589C6b3ca184111"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x5faf16A85028BE138A7178B222DeC98092FEEF97"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xF965237c6b3f89f8C62B45b94097899E3562A830"),
        erc1155: (0, types_1.toEVMAddress)("0xd37DC0CD86Dfa9B2B57CD7DFA8B6AA0092a9517d"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x4200000000000000000000000000000000000006"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
