// @rarible/protocol-ethereum-sdk/build/config/base-sepolia.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.baseSepoliaConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.baseSepoliaConfig = {
    basePath: "https://testnet-base-api.rarible.org",
    chainId: 84532,
    environment: "testnet",
    blockchain: api_client_1.Blockchain.BASE,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x0fE65B68Eb627c21EAF3cfe8183C4F946F3d48BD"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x4fEB488209d2A0A71fEef28E5fA306F15b2D5FEa"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x4f4cC63D7f2bC894078d41f284453062842Afa46"),
        erc20: (0, types_1.toEVMAddress)("0xBc57D6e50fa760A01536A7c2EAEDD6fC9b2A4f9A"),
        erc721Lazy: (0, types_1.toEVMAddress)("0xF1cb795B7eA59a9213a790f868104c11a14690Fa"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0xa90e536A492aef0d57e6d295FA17687c3ca93347"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xB020bA7fcF43DCc59eF0103624BD6FADE66d105E"),
        erc1155: (0, types_1.toEVMAddress)("0x166F6180170f438Ddc38050a2B708d38c0890956"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x4200000000000000000000000000000000000006"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
