// @rarible/protocol-ethereum-sdk/build/config/fivire.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fivireConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const utils_1 = require("./utils");
exports.fivireConfig = (0, utils_1.extendConfig)({
    basePath: "https://fivire-api.rarible.org",
    chainId: 995,
    environment: "production",
    blockchain: api_client_1.Blockchain.FIVIRE,
    exchange: {
        v2: (0, types_1.toEVMAddress)("0x9b761A2C45daEd76Dfbcfd52d22cB930a0b41186"),
        wrapper: (0, types_1.toEVMAddress)("0x55A921a57f7F15dF2f229Ab9889506Ca89310800"),
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x57559104c0e4fbc578fF4251b63eA0903d7CAe32"),
        erc20: (0, types_1.toEVMAddress)("0x4217a346C8b48731641327b65bb6F6d3243d64e2"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x1CC22424f2B84791cb99c141A68CD2a44Cf35398"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0xBCE7d7fbA750B1E9e0511C67b1F38C07EbfEFE63"),
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xD9F3BfeD52ec008A13cF08C7382a917Eb364Cc32"),
        erc1155: (0, types_1.toEVMAddress)("0xBFb17500344bA3475d46091F5c8f1e33B31ed909"),
    },
    weth: types_1.EVM_ZERO_ADDRESS,
});
