// @rarible/protocol-ethereum-sdk/build/config/kroma.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.kromaConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.kromaConfig = {
    basePath: "https://kroma-api.rarible.org",
    chainId: 255,
    environment: "production",
    blockchain: api_client_1.Blockchain.KROMA,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x418f1b76448866CF072dd14d092138190CcdC9aF"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x69Be0b6f5BB5e9F8DfAA1562F06427142fb0a10a"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x30fc6eed1d302F5f5C4a8aa58047d1a730b3Cc91"),
        erc20: (0, types_1.toEVMAddress)("0x6C65a3C3AA67b126e43F86DA85775E0F5e9743F7"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x9f4CE147beF144a571c1372cd6e1DEB148742027"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x8d5D3e1e6609A798B84160Ec9CC25198B9e4F177"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x24562E94332f06F1F0aEa79cAf283088F9e1B8A8"),
        erc1155: (0, types_1.toEVMAddress)("0xbd426Ce20Ac84ad3a6FA6FF50cfe5fB325AB0e5d"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x4200000000000000000000000000000000000001"),
    auction: types_1.EVM_ZERO_ADDRESS,
};
