// @rarible/protocol-ethereum-sdk/build/config/lisk.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.liskMainnetConfig = void 0;
const api_client_1 = require("@rarible/api-client");
const types_1 = require("@rarible/types");
const utils_1 = require("./utils");
exports.liskMainnetConfig = (0, utils_1.extendConfig)({
    basePath: "https://lisk-api.rarible.org",
    chainId: 1135,
    environment: "production",
    blockchain: api_client_1.Blockchain.LISK,
    exchange: {
        v2: (0, types_1.toEVMAddress)("0x5faf16A85028BE138A7178B222DeC98092FEEF97"),
        wrapper: (0, types_1.toEVMAddress)("0xcD6A173F1C244C3d9b9bc2434582e54B87739F08"),
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0xF65eF65a95821A16E02973b1C2200FA58898e3c0"),
        erc20: (0, types_1.toEVMAddress)("0x248B46BEB66b3078D771a9E7E5a0a0216d0d07ba"),
        erc721Lazy: (0, types_1.toEVMAddress)("0xce4bf732f53A76C463aE8822be858017b02779c8"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x1CC22424f2B84791cb99c141A68CD2a44Cf35398"),
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x69Be0b6f5BB5e9F8DfAA1562F06427142fb0a10a"),
        erc1155: (0, types_1.toEVMAddress)("0xf5c9643bE5C6925F2272ecA95De16e002D6fC83C"),
    },
    weth: (0, types_1.toEVMAddress)("0x4200000000000000000000000000000000000006"),
});
