// @rarible/protocol-ethereum-sdk/build/config/moonbeam.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.moonbeamMainnetConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.moonbeamMainnetConfig = {
    basePath: "https://moonbeam-api.rarible.org",
    chainId: 1284,
    environment: "production",
    blockchain: api_client_1.Blockchain.MOONBEAM,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x30fc6eed1d302F5f5C4a8aa58047d1a730b3Cc91"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0xEA26e060cCc11C840e6107cfca0B41c45Ce6a5a2"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x57559104c0e4fbc578fF4251b63eA0903d7CAe32"),
        erc20: (0, types_1.toEVMAddress)("0x4217a346C8b48731641327b65bb6F6d3243d64e2"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x13b05523634ABb96E6017Da71b7698CAecDf50b2"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x339e61eb644A29B134D7fD3fA589C6b3ca184111"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x726f5DD8a1fFAa349bD57501Ba760AB5A03e91Ff"),
        erc1155: (0, types_1.toEVMAddress)("0x55A921a57f7F15dF2f229Ab9889506Ca89310800"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0xab3f0245b83feb11d15aaffefd7ad465a59817ed"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
