// @rarible/protocol-ethereum-sdk/build/config/mumbai.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mumbaiConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.mumbaiConfig = {
    basePath: "https://testnet-polygon-api.rarible.org",
    chainId: 80001,
    environment: "testnet",
    blockchain: api_client_1.Blockchain.POLYGON,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x2Fc743F5419637B93dDAC159715B902186300041"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x042221c65AC6925fbc478dB5B746183f72377526"),
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x02e21199D043dab90248f79d6A8d0c36832734B0"),
        erc20: (0, types_1.toEVMAddress)("0xCA90c2E7d3c41EF220888B0038849BA1e67688bC"),
        erc721Lazy: (0, types_1.toEVMAddress)("0xA8628124C255171cF356F0E0204E2D19CA89F636"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x8fd27f771bf72C1F8c8b78FCdcf323C0C34f030b"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xa85180a21786bA65b0778bE1cb5CBA5E5c6cD21d"),
        erc1155: (0, types_1.toEVMAddress)("0xAa9CD5834E0009902EeAA3FEfAc6A160e9A096b4"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x9c3c9283d3e44854697cd22d3faa240cfb032889"),
    auction: types_1.EVM_ZERO_ADDRESS,
};
