// @rarible/protocol-ethereum-sdk/build/config/polygon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.polygonConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.polygonConfig = {
    basePath: "https://polygon-api.rarible.org",
    chainId: 137,
    environment: "production",
    blockchain: api_client_1.Blockchain.POLYGON,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x12b3897a36fDB436ddE2788C06Eff0ffD997066e"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0xe90da87Ec96DF89590A5CD00c0183E69a36330a9"),
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0xd47e14DD9b98411754f722B4c4074e14752Ada7C"),
        erc20: (0, types_1.toEVMAddress)("0x49b4e47079d9b733B2227fa15f0762dBF707B263"),
        erc721Lazy: (0, types_1.toEVMAddress)("0xDD28328257a2Cce3204332C747Cc350153937A1D"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x0E63021A7597B254484b7F99dDD9b319591350B6"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x16911a36a56f828f17632cD4915614Dd5c7a45e0"),
        erc1155: (0, types_1.toEVMAddress)("0xF46e8e6fA0F048DdD76F8c6982eBD059796298B8"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x0d500b1d8e8ef31e21c99d1db9a6444d3adf1270"),
    auction: types_1.EVM_ZERO_ADDRESS,
};
