// @rarible/protocol-ethereum-sdk/build/config/rari.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rariMainnetConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.rariMainnetConfig = {
    basePath: "https://rari-api.rarible.org",
    chainId: 1380012617,
    environment: "production",
    blockchain: api_client_1.Blockchain.RARI,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x10CCBf49617ECB7A8262065853D6C93Ad42C3C2C"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0xd37DC0CD86Dfa9B2B57CD7DFA8B6AA0092a9517d"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0xce4bf732f53A76C463aE8822be858017b02779c8"),
        erc20: (0, types_1.toEVMAddress)("0x1CC22424f2B84791cb99c141A68CD2a44Cf35398"),
        erc721Lazy: (0, types_1.toEVMAddress)("0xBCE7d7fbA750B1E9e0511C67b1F38C07EbfEFE63"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x30fc6eed1d302F5f5C4a8aa58047d1a730b3Cc91"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xdA12E4Ab1d731F29bF4Bff8f971579D95f8DDD07"),
        erc1155: (0, types_1.toEVMAddress)("0xEA26e060cCc11C840e6107cfca0B41c45Ce6a5a2"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0xf037540e51D71b2D2B1120e8432bA49F29EDFBD0"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
