// @rarible/protocol-ethereum-sdk/build/config/sei-arctic-1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.seiArctic1Config = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.seiArctic1Config = {
    basePath: "https://testnet-sei-api.rarible.org",
    chainId: 713715,
    environment: "testnet",
    blockchain: api_client_1.Blockchain.SEI,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x5b96ddd3245aaC0cd97cFf6717985F3b9B01df3B"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x0D7147461cef9Ce51B8ba63560Dc53f96E90638B"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0xA5707153b8AF346e17AF765235B054136eCf99EC"),
        erc20: (0, types_1.toEVMAddress)("0xA3d49c7e2c845b792e422696FE0D9ef17a509731"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x74eBEEB6FBCf94f748f5999E14aCf3642A38e813"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0xd60D801E1E76a44f8A2E728F6d3a760626aa2cf2"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xE5D43624116007CD79D17fB7136672A49fd33CE0"),
        erc1155: (0, types_1.toEVMAddress)("0x6E6d179b1DCEaad5483edaCF907bf619FB35C7Eb"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x57ee725beeb991c70c53f9642f36755ec6eb2139"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
