// @rarible/protocol-ethereum-sdk/build/config/shape.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.shapeMainnetConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const utils_1 = require("./utils");
exports.shapeMainnetConfig = (0, utils_1.extendConfig)({
    basePath: "https://shape-api.rarible.org",
    chainId: 360,
    environment: "production",
    blockchain: api_client_1.Blockchain.SHAPE,
    exchange: {
        v2: (0, types_1.toEVMAddress)("0x30fc6eed1d302F5f5C4a8aa58047d1a730b3Cc91"),
        wrapper: (0, types_1.toEVMAddress)("0xBFb17500344bA3475d46091F5c8f1e33B31ed909"),
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x57559104c0e4fbc578fF4251b63eA0903d7CAe32"),
        erc20: (0, types_1.toEVMAddress)("0x4217a346C8b48731641327b65bb6F6d3243d64e2"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x13b05523634ABb96E6017Da71b7698CAecDf50b2"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x339e61eb644A29B134D7fD3fA589C6b3ca184111"),
    },
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x726f5DD8a1fFAa349bD57501Ba760AB5A03e91Ff"),
        erc1155: (0, types_1.toEVMAddress)("0xF965237c6b3f89f8C62B45b94097899E3562A830"),
    },
    weth: (0, types_1.toEVMAddress)("0x4200000000000000000000000000000000000006"),
});
