// @rarible/protocol-ethereum-sdk/build/config/testnet-mantle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mantleTestnetConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.mantleTestnetConfig = {
    basePath: "https://testnet-mantle-api.rarible.org",
    chainId: 5003,
    environment: "testnet",
    blockchain: api_client_1.Blockchain.MANTLE,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x61512179F6a16bEC0D259d8010CC0485CE363868"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0xc798B273FaF23932Cf11177402C10C9b44D30Da2"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x00C74eD067Cea48F1D6F7D00aBABa3C1D5B2598b"),
        erc20: (0, types_1.toEVMAddress)("0xA094E566b61b3c2D88ACf7Cc15e3Dd0FA83F32af"),
        erc721Lazy: (0, types_1.toEVMAddress)("0xB8863180CAC2d0Ab665e5968C0De25298A1D8CEe"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0xa2eEBb837aEF89369Ad117568d75348e6174520e"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xe10605b2026884aCc669C2A9Cd4A5ec5f5FFf494"),
        erc1155: (0, types_1.toEVMAddress)("0x57B3f3b79F64c475a37E6c493BAA8E6E7C5F748e"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0xdEAddEaDdeadDEadDEADDEAddEADDEAddead1111"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
