// @rarible/protocol-ethereum-sdk/build/config/testnet-zksync.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zkSyncTestnetConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.zkSyncTestnetConfig = {
    basePath: "https://testnet-zksync-api.rarible.org",
    chainId: 300,
    environment: "testnet",
    blockchain: api_client_1.Blockchain.ZKSYNC,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x99bD7BA01f9872f034a35DC4bC737cFaaaC11D63"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x0148b11891C0E30Fb36a6D646E04C7bebE7969c8"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x85a26E6D52239817570Ff643bA09E3AA5393A805"),
        erc20: (0, types_1.toEVMAddress)("0x43b9B5221F513031acC62dc8B9788E608B293baD"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x11983886da3c379E507A874649C96D7EEd086c32"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x117c152C992e8c344Ce5a84100130cd87eF6bAE6"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x2abC8bD09c681702e2297F11fd759d9014664B33"),
        erc1155: (0, types_1.toEVMAddress)("0xEda115D72EF1Dc96d0297278D1D241821F999927"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0xdf09a97A1CF809C335616c21c3a0EA4780F96514"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
