// @rarible/protocol-ethereum-sdk/build/config/testnet-lightlink.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testnetLightlinkConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.testnetLightlinkConfig = {
    basePath: "https://testnet-lightlink-api.rarible.org",
    chainId: 1891,
    environment: "testnet",
    blockchain: api_client_1.Blockchain.LIGHTLINK,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x2E015B0474364757d2cc8e28897DCBCdEE07e340"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: types_1.EVM_ZERO_ADDRESS,
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x166F6180170f438Ddc38050a2B708d38c0890956"),
        erc20: (0, types_1.toEVMAddress)("0x7d47126a2600E22eab9eD6CF0e515678727779A6"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x98C2d878064dCD20489214cf0866f972f91784D0"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x12B372153249F006F756d0668fCDBD8fbD8b0a15"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xfeC0F8d936B9cBa92a332bCB06dC7DF4DdE0c253"),
        erc1155: (0, types_1.toEVMAddress)("0x63e3297a90B4101d0a4Bb8EbEFDF3D47C8d4D4Ac"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0xF42991f02C07AB66cFEa282E7E482382aEB85461"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
