// @rarible/tezos-sdk/dist/order/fill-order.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fill_order = exports.get_rarible_legacy_buy_transaction = exports.match_order_to_struct = void 0;
var tezos_common_1 = require("@rarible/tezos-common");
var utils_1 = require("./utils");
var invert_order_1 = require("./invert-order");
var get_make_fee_1 = require("./get-make-fee");
var add_fee_1 = require("./add-fee");
var sign_order_1 = require("./sign-order");
var bignumber_js_1 = require("bignumber.js");
function get_make_asset(provider, order, amount, edpk) {
    return __awaiter(this, void 0, void 0, function () {
        var inverted, make_fee;
        return __generator(this, function (_a) {
            inverted = (0, invert_order_1.invert_order)(order, amount, edpk);
            make_fee = (0, get_make_fee_1.get_make_fee)(provider.config.fees, inverted);
            return [2 /*return*/, (0, add_fee_1.add_fee)(provider, inverted.make, make_fee)];
        });
    });
}
function get_real_value(provider, order) {
    return __awaiter(this, void 0, void 0, function () {
        var fee, make;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    fee = (0, get_make_fee_1.get_make_fee)(provider.config.fees, order);
                    return [4 /*yield*/, (0, add_fee_1.add_fee)(provider, order.make, fee)];
                case 1:
                    make = _a.sent();
                    return [2 /*return*/, make.value];
            }
        });
    });
}
function match_order_to_struct(p, left, right) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, _b;
        var _c, _d, _e;
        return __generator(this, function (_f) {
            switch (_f.label) {
                case 0:
                    _c = {
                        prim: "Pair"
                    };
                    return [4 /*yield*/, (0, sign_order_1.order_to_struct)(p, left)];
                case 1:
                    _a = [
                        _f.sent()
                    ];
                    _d = {
                        prim: "Pair"
                    };
                    _b = [(left.signature) ? (0, sign_order_1.some_struct)({ string: left.signature }) : (0, sign_order_1.none_struct)()];
                    _e = {
                        prim: "Pair"
                    };
                    return [4 /*yield*/, (0, sign_order_1.order_to_struct)(p, right)];
                case 2: return [2 /*return*/, (_c.args = _a.concat([
                        (_d.args = _b.concat([
                            (_e.args = [
                                _f.sent(),
                                (right.signature) ? (0, sign_order_1.some_struct)({ string: right.signature }) : (0, sign_order_1.none_struct)()
                            ],
                                _e)
                        ]),
                            _d)
                    ]),
                        _c)];
            }
        });
    });
}
exports.match_order_to_struct = match_order_to_struct;
function get_rarible_legacy_buy_transaction(provider, left, request) {
    return __awaiter(this, void 0, void 0, function () {
        var pk, _a, make, right, amount_to_approve, amount, asset, decimals, asset, decimals, arg_approve, _b, _c, _d, args, parameter;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    if (!(request.edpk)) return [3 /*break*/, 1];
                    _a = request.edpk;
                    return [3 /*break*/, 3];
                case 1: return [4 /*yield*/, (0, tezos_common_1.get_public_key)(provider)];
                case 2:
                    _a = _e.sent();
                    _e.label = 3;
                case 3:
                    pk = _a;
                    if (!pk)
                        throw new Error("cannot get public key");
                    return [4 /*yield*/, get_make_asset(provider, left, request.amount, pk)];
                case 4:
                    make = _e.sent();
                    right = __assign(__assign({}, (0, invert_order_1.invert_order)(left, request.amount, pk)), { data: __assign(__assign({}, left.data), { payouts: request.payouts || [], origin_fees: request.origin_fees || [] }) });
                    return [4 /*yield*/, (0, utils_1.fill_offchain_royalties)(provider, right)];
                case 5:
                    right = _e.sent();
                    amount_to_approve = undefined;
                    amount = undefined;
                    if (!((left.make.asset_type.asset_class === "XTZ" || left.make.asset_type.asset_class === "FT") && left.salt == '0')) return [3 /*break*/, 9];
                    return [4 /*yield*/, get_real_value(provider, left)];
                case 6:
                    amount = _e.sent();
                    if (!(left.make.asset_type.asset_class === "FT")) return [3 /*break*/, 8];
                    asset = left.make.asset_type;
                    return [4 /*yield*/, (0, tezos_common_1.get_decimals)(provider.config, asset.contract, asset.token_id)];
                case 7:
                    decimals = _e.sent();
                    amount_to_approve = amount.times(new bignumber_js_1.default("10").pow(decimals));
                    _e.label = 8;
                case 8: return [3 /*break*/, 12];
                case 9:
                    if (!((right.make.asset_type.asset_class === "XTZ" || right.make.asset_type.asset_class === "FT") && right.salt == '0')) return [3 /*break*/, 12];
                    return [4 /*yield*/, get_real_value(provider, right)];
                case 10:
                    amount = _e.sent();
                    if (!(right.make.asset_type.asset_class === "FT")) return [3 /*break*/, 12];
                    asset = right.make.asset_type;
                    return [4 /*yield*/, (0, tezos_common_1.get_decimals)(provider.config, asset.contract, asset.token_id)];
                case 11:
                    decimals = _e.sent();
                    amount_to_approve = amount.times(new bignumber_js_1.default("10").pow(decimals));
                    _e.label = 12;
                case 12:
                    if (!(make.asset_type.asset_class != "XTZ")) return [3 /*break*/, 15];
                    _c = tezos_common_1.approve_arg;
                    _d = [provider];
                    return [4 /*yield*/, (0, tezos_common_1.get_address)(provider)];
                case 13: return [4 /*yield*/, _c.apply(void 0, _d.concat([_e.sent(), make, amount_to_approve, request.use_all]))];
                case 14:
                    _b = _e.sent();
                    return [3 /*break*/, 16];
                case 15:
                    _b = undefined;
                    _e.label = 16;
                case 16:
                    arg_approve = _b;
                    args = (arg_approve) ? [arg_approve] : [];
                    return [4 /*yield*/, match_order_to_struct(provider, left, right)];
                case 17:
                    parameter = _e.sent();
                    args = args.concat({
                        destination: provider.config.exchange, entrypoint: "match_orders",
                        parameter: parameter,
                        amount: amount
                    });
                    return [2 /*return*/, args];
            }
        });
    });
}
exports.get_rarible_legacy_buy_transaction = get_rarible_legacy_buy_transaction;
function fill_order(provider, left, request) {
    return __awaiter(this, void 0, void 0, function () {
        var args;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, get_rarible_legacy_buy_transaction(provider, left, request)];
                case 1:
                    args = _a.sent();
                    return [2 /*return*/, (0, tezos_common_1.send_batch)(provider, args)];
            }
        });
    });
}
exports.fill_order = fill_order;
