// @rarible/immutable-sdk/build/apis/balance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImxBalanceControllerApi = void 0;
const tslib_1 = require("tslib");
const ethereum_api_client_1 = require("@rarible/ethereum-api-client");
class ImxBalanceControllerApi extends ethereum_api_client_1.BaseAPI {
    getAllBalancesRaw(requestParameters) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (requestParameters.ownerAddress === null || requestParameters.ownerAddress === undefined) {
                throw new ethereum_api_client_1.RequiredError("ownerAddress", "Required parameter requestParameters.ownerAddress was null or undefined when calling getAllBalances.");
            }
            const queryParameters = {};
            const headerParameters = {};
            const response = yield this.request({
                path: "/balances/{address}".replace(`{${"address"}}`, encodeURIComponent(String(requestParameters.ownerAddress))),
                method: "GET",
                headers: headerParameters,
                query: queryParameters,
            });
            return new ethereum_api_client_1.JSONApiResponse(response, jsonValue => jsonValue);
        });
    }
    /**
     */
    getAllBalances(requestParameters) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const response = yield this.getAllBalancesRaw(requestParameters);
            return yield response.value();
        });
    }
}
exports.ImxBalanceControllerApi = ImxBalanceControllerApi;
