// @rarible/protocol-ethereum-sdk/build/order/contracts/sudoswap-pair.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUDOSWAP_PAIR_ABI = exports.createSudoswapPairContract = void 0;
function createSudoswapPairContract(ethereum, address) {
    return ethereum.createContract(exports.SUDOSWAP_PAIR_ABI, address);
}
exports.createSudoswapPairContract = createSudoswapPairContract;
exports.SUDOSWAP_PAIR_ABI = [
    {
        inputs: [
            {
                internalType: "uint256",
                name: "numNFTs",
                type: "uint256",
            },
        ],
        name: "getBuyNFTQuote",
        outputs: [
            {
                internalType: "enum CurveErrorCodes.Error",
                name: "error",
                type: "uint8",
            },
            {
                internalType: "uint256",
                name: "newSpotPrice",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "newDelta",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "inputAmount",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "protocolFee",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
