// @rarible/protocol-ethereum-sdk/build/config/arbitrum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.arbitrumConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.arbitrumConfig = {
    basePath: "https://arbitrum-api.rarible.org",
    chainId: 42161,
    environment: "production",
    blockchain: api_client_1.Blockchain.ARBITRUM,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x07b637739CAd9A5f0c487219B283a52717E69978"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x13b05523634ABb96E6017Da71b7698CAecDf50b2"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x49b4e47079d9b733B2227fa15f0762dBF707B263"),
        erc20: (0, types_1.toEVMAddress)("0xDD28328257a2Cce3204332C747Cc350153937A1D"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x0E63021A7597B254484b7F99dDD9b319591350B6"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x1Bea70C8c949b3b8d1188cb738432B121B83C4b5"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xdA12E4Ab1d731F29bF4Bff8f971579D95f8DDD07"),
        erc1155: (0, types_1.toEVMAddress)("0xEA26e060cCc11C840e6107cfca0B41c45Ce6a5a2"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x82aF49447D8a07e3bd95BD0d56f35241523fBab1"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
