// @rarible/protocol-ethereum-sdk/build/config/celo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.celoConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.celoConfig = {
    basePath: "https://celo-api.rarible.org",
    environment: "production",
    chainId: 42220,
    blockchain: api_client_1.Blockchain.CELO,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x5faf16A85028BE138A7178B222DeC98092FEEF97"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0xBFb17500344bA3475d46091F5c8f1e33B31ed909"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0xF65eF65a95821A16E02973b1C2200FA58898e3c0"),
        erc20: (0, types_1.toEVMAddress)("0x248B46BEB66b3078D771a9E7E5a0a0216d0d07ba"),
        erc721Lazy: (0, types_1.toEVMAddress)("0xce4bf732f53A76C463aE8822be858017b02779c8"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x1CC22424f2B84791cb99c141A68CD2a44Cf35398"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xC699FB932c1bD7235C7ED19388f26A2428224AED"),
        erc1155: (0, types_1.toEVMAddress)("0xdA12E4Ab1d731F29bF4Bff8f971579D95f8DDD07"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: types_1.EVM_ZERO_ADDRESS,
    auction: types_1.EVM_ZERO_ADDRESS,
};
