// @rarible/protocol-ethereum-sdk/build/config/chilliz.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chilizConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.chilizConfig = {
    basePath: "https://chiliz-api.rarible.org",
    chainId: 88888,
    environment: "production",
    blockchain: api_client_1.Blockchain.CHILIZ,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0xdA12E4Ab1d731F29bF4Bff8f971579D95f8DDD07"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0xEe07D8603F4A8A72B523314D0D473a6Ce93354Fe"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x8d5D3e1e6609A798B84160Ec9CC25198B9e4F177"),
        erc20: (0, types_1.toEVMAddress)("0x0B7Acd053BC236fc95537e2aDD37C22968b48C1b"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x519D6A81b2894FC5e0F2a8B900F6f5CdE1132dBB"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x726f5DD8a1fFAa349bD57501Ba760AB5A03e91Ff"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: types_1.EVM_ZERO_ADDRESS,
        erc1155: types_1.EVM_ZERO_ADDRESS,
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x677F7e16C7Dd57be1D4C8aD1244883214953DC47"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
