// @rarible/protocol-ethereum-sdk/build/config/dev.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.devEthereumConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.devEthereumConfig = {
    basePath: "https://dev-ethereum-api.rarible.org",
    chainId: 300500,
    environment: "dev",
    blockchain: api_client_1.Blockchain.ETHEREUM,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x3fB287d1Da10a10A87b613dED57230964e546719"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x8edcb67dd394AFfe535BfedF8B2ed191Be8BCB36"),
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x0b6F1b558b3808EA1B10e5ac29cA82c234C7ca4c"),
        erc20: (0, types_1.toEVMAddress)("0xa721f321f2C3838e6812b1c8b1693e3B1f6a38Bc"),
        erc721Lazy: (0, types_1.toEVMAddress)("0xc6f33b62A94939E52E1b074c4aC1A801B869fDB2"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x3586d3E6CDaE98d5F0eEaB737977Bc78406Da2BD"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x6aABb267a1c440CfB5C200Ebcd078Efa9249492A"),
        erc1155: (0, types_1.toEVMAddress)("0x8283Ffd0F535E1103C3599D2d00b85815774A896"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: (0, types_1.toEVMAddress)("0x6972347e66A32F40ef3c012615C13cB88Bf681cc"),
        pairRouter: (0, types_1.toEVMAddress)("0x319c4Bd373d3F16697d630153F5a2d526047FD8C"),
    },
    weth: (0, types_1.toEVMAddress)("0x3554BA6cb4862C7CB2463f461deF81FA4A8f8E3C"),
    auction: types_1.EVM_ZERO_ADDRESS,
};
