// @rarible/protocol-ethereum-sdk/build/config/mantle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mantleConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.mantleConfig = {
    basePath: "https://mantle-api.rarible.org",
    chainId: 5000,
    environment: "production",
    blockchain: api_client_1.Blockchain.MANTLE,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x0e7B24d73e45B639A5cF674C5f2Bb02930716f87"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: types_1.EVM_ZERO_ADDRESS,
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0xA3AaA33E13Bd42fE5cBDefC72fB0a0888cfB44C0"),
        erc20: (0, types_1.toEVMAddress)("0x5274ac9507b20aC14e215B098479bd69733fA98A"),
        erc721Lazy: (0, types_1.toEVMAddress)("0xc0C8d44A78605E4C221C9506DA737bB2A5dfd537"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x2047f99EFa18009ceA518AC99cEE8e2151D53eDc"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x465d62a669E98517e08e4E3D809A28FAF3DfbAE1"),
        erc1155: (0, types_1.toEVMAddress)("0x16911a36a56f828f17632cD4915614Dd5c7a45e0"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0xdeaddeaddeaddeaddeaddeaddeaddeaddead1111"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
