// @rarible/protocol-ethereum-sdk/build/config/oasys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.oasysConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.oasysConfig = {
    basePath: "https://oasys-api.rarible.org",
    chainId: 248,
    environment: "production",
    blockchain: api_client_1.Blockchain.OASYS,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0xfD79EF502b5b97E86804A9359fE5Ba48AB3658c5"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0xa3af07132CeA0573880B25d4f89653ab87774c4c"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x9ABc3Ce3a35873274dE3B8b6c15D3BD1F6F2Ec8D"),
        erc20: (0, types_1.toEVMAddress)("0x15FCB776c6A0E050FF0e9b423a0EAD1F0dfE1196"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x56b666895EFab1fb4Fa29298F390c380126d581c"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x7698983DF7cd2A57aCDF97656aebFf64398A60a7"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xEe962828d39ec46962B883bfFeBa0721499d435C"),
        erc1155: (0, types_1.toEVMAddress)("0xD9Cb91C9b6Aa00e75ef856557CA86371599be3e8"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x5200000000000000000000000000000000000001"),
    auction: types_1.EVM_ZERO_ADDRESS,
};
