// @rarible/protocol-ethereum-sdk/build/config/polygon-amoy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.polygonAmoyConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.polygonAmoyConfig = {
    basePath: "https://testnet-polygon-api.rarible.org",
    chainId: 80002,
    environment: "testnet",
    blockchain: api_client_1.Blockchain.POLYGON,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x3e52D660b69d1bDacb6C513cE085D924F5Cb9c77"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x2FCE8435F0455eDc702199741411dbcD1B7606cA"),
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0xA094E566b61b3c2D88ACf7Cc15e3Dd0FA83F32af"),
        erc20: (0, types_1.toEVMAddress)("0xB8863180CAC2d0Ab665e5968C0De25298A1D8CEe"),
        erc721Lazy: (0, types_1.toEVMAddress)("0xa2eEBb837aEF89369Ad117568d75348e6174520e"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0xC5BBd75789bD007784A0046094d19aCeA1A79eB1"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x18a2553ef1aaE12d9cd158821319e26A62feE90E"),
        erc1155: (0, types_1.toEVMAddress)("0xc9eB416CDb5cc2aFC09bb75393AEc6dBA4E5C84a"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x0ae690aad8663aab12a671a6a0d74242332de85f"),
    auction: types_1.EVM_ZERO_ADDRESS,
};
