// @rarible/protocol-ethereum-sdk/build/config/saakuru.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saakuruConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.saakuruConfig = {
    basePath: "https://saakuru-api.rarible.org",
    chainId: 7225878,
    environment: "production",
    blockchain: api_client_1.Blockchain.SAAKURU,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0xEbAB63727F1E9163B34CD862388cE33Bd5Be3199"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: types_1.EVM_ZERO_ADDRESS,
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x39C41C685D26d0219D4717235C2bf6E56055fa4d"),
        erc20: (0, types_1.toEVMAddress)("0xbd426Ce20Ac84ad3a6FA6FF50cfe5fB325AB0e5d"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x9847154Ec2d4009c2F067926d554F0d3986e1f64"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0xdC612825cc0Ab6e2CCDf0Cb4E45D4C68D69E21b4"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x9f283cA6a29864F32ec44879Fc4B509ab3277eeD"),
        erc1155: (0, types_1.toEVMAddress)("0x96992D63fE4317C7a99f0Ec2a2C0CcbC640fD29a"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x557a526472372f1f222ecc6af8818c1e6e78a85f"),
    auction: types_1.EVM_ZERO_ADDRESS,
};
