// @rarible/protocol-ethereum-sdk/build/config/sei-pacific-1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.seiPacific1Config = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.seiPacific1Config = {
    basePath: "https://sei-api.rarible.org",
    chainId: 1329,
    environment: "production",
    blockchain: api_client_1.Blockchain.SEI,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x42b8DB7aCB54B4f8690C7379Ff2Befb6caf67478"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x4Da5504b5FD5C5073906478B2e9B66d278aB4D2C"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x8dBEcA8fA7ed5424f8b8dD5945dDC62393D0a642"),
        erc20: (0, types_1.toEVMAddress)("0xb1199ECB7bdB9eE082b9535A6c08c912914CaBAC"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x9f1e78A81fF684F034efBbd97cCE508dF19B4210"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x56A3A3D3DeC25526eb54C771B15159443672602A"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x4fA35DD15183a320254fE36aE1CC7067dDAC112f"),
        erc1155: (0, types_1.toEVMAddress)("0xa008fe83286382B4D56D92A07C4E602d58AF62F1"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0xe30fedd158a2e3b13e9badaeabafc5516e95e8c7"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
