// @rarible/protocol-ethereum-sdk/build/config/testnet-chiliz.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chilizTestnetConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.chilizTestnetConfig = {
    basePath: "https://testnet-chiliz-api.rarible.org",
    chainId: 88882,
    environment: "testnet",
    blockchain: api_client_1.Blockchain.CHILIZ,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x4c27bE9fE53227194Ff259D8906A2A1b0479A3AA"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x1fD75d68F0D0F66383F011D282890BDACE221Dc2"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0x3d1C134ece4CFA4c44AE5D37f74dEeccBceC9031"),
        erc20: (0, types_1.toEVMAddress)("0x8A42da3cfd53ff38E6551cc3a05F536428DaaE34"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x615fdFC73edB58d9ef09574B5284E6E6362F7f6D"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x40785643bdD364A21aeE1d138E026e8914c98572"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xAeEfB55eD03eC5a25Fc4C84354b6C8c65Df963EA"),
        erc1155: (0, types_1.toEVMAddress)("0x7c512F690E89CF01deb04Bc68af95b1A5f7A2504"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x678c34581db0a7808d0aC669d7025f1408C9a3C6"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
