// @rarible/protocol-ethereum-sdk/build/config/testnet-celo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.celoTestnetConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.celoTestnetConfig = {
    basePath: "https://testnet-celo-api.rarible.org",
    environment: "testnet",
    chainId: 44787,
    blockchain: api_client_1.Blockchain.CELO,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0xB4D34a10921347877B0AA7A9DB347871b20b19F5"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0x7D63585bEF6FA1D49d70558FF0616C99480FFA0F"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0xf1dCB818F494Fb63358510b6d05Cc50096B8F06c"),
        erc20: (0, types_1.toEVMAddress)("0xB02f8F8F3527e5b2C7dB72B7eE1Af244fA8B3BAE"),
        erc721Lazy: (0, types_1.toEVMAddress)("0xE3Baf1b17335bbf3AC3C2cFCe95eC1bfC463d0c8"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x7Eabe83e0F99B6bf24Ec3F50994B972DC38D11dF"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0x48838abEAE900a2FC9fC4eC95a47F29a6c1B7647"),
        erc1155: (0, types_1.toEVMAddress)("0xC1e685AF493CcC473F22664151947CDA56Fae0A1"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: types_1.EVM_ZERO_ADDRESS,
    auction: types_1.EVM_ZERO_ADDRESS,
};
