// @rarible/protocol-ethereum-sdk/build/config/zksync.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zkSyncConfig = void 0;
const types_1 = require("@rarible/types");
const api_client_1 = require("@rarible/api-client");
const id_1 = require("../common/id");
const common_1 = require("./common");
exports.zkSyncConfig = {
    basePath: "https://zksync-api.rarible.org",
    chainId: 324,
    environment: "production",
    blockchain: api_client_1.Blockchain.ZKSYNC,
    exchange: {
        v1: types_1.EVM_ZERO_ADDRESS,
        v2: (0, types_1.toEVMAddress)("0x5E0BbEd68e1b47C94a396226D8AC10DDe242e77c"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        wrapper: (0, types_1.toEVMAddress)("0xEf3b8F0B7EE374F5F79BE4D43E8cbB4A7952f274"),
        looksrare: types_1.EVM_ZERO_ADDRESS,
        looksrareV2: types_1.EVM_ZERO_ADDRESS,
        x2y2: types_1.EVM_ZERO_ADDRESS,
    },
    transferProxies: {
        nft: (0, types_1.toEVMAddress)("0xaf63698293A4c7d235CCf6F809C348D641C0bd62"),
        erc20: (0, types_1.toEVMAddress)("0xb5986bB35a6b53cb4764951Ad83cA12fa5a51C64"),
        erc721Lazy: (0, types_1.toEVMAddress)("0x463651f1620E411426E7eB70c3D2029106F2B6E0"),
        erc1155Lazy: (0, types_1.toEVMAddress)("0x99e3d07C2fA7d9566bAA34e84B9DD5b8fB98961a"),
        openseaV1: types_1.EVM_ZERO_ADDRESS,
        cryptoPunks: types_1.EVM_ZERO_ADDRESS,
    },
    feeConfigUrl: common_1.FEE_CONFIG_URL,
    openSea: {
        metadata: (0, id_1.id32)("RARIBLE"),
        proxyRegistry: types_1.EVM_ZERO_ADDRESS,
        merkleValidator: types_1.EVM_ZERO_ADDRESS,
    },
    factories: {
        erc721: (0, types_1.toEVMAddress)("0xB38F451e6Cc0Ad0e7a31C6Ec5648177Ba248eE9B"),
        erc1155: (0, types_1.toEVMAddress)("0x196e1D96e73c805ee39C766435A81fb235510939"),
    },
    cryptoPunks: {
        marketContract: types_1.EVM_ZERO_ADDRESS,
        wrapperContract: types_1.EVM_ZERO_ADDRESS,
    },
    sudoswap: {
        pairFactory: types_1.EVM_ZERO_ADDRESS,
        pairRouter: types_1.EVM_ZERO_ADDRESS,
    },
    weth: (0, types_1.toEVMAddress)("0x8Ebe4A94740515945ad826238Fc4D56c6B8b0e60"),
    auction: types_1.EVM_ZERO_ADDRESS,
    looksrareOrderValidatorV2: types_1.EVM_ZERO_ADDRESS,
};
